{"cells":[{"metadata":{},"cell_type":"markdown","source":"<h1 style=\"background-color:#DC143C; font-family:'Brush Script MT',cursive;color:white;font-size:200%; text-align:center;border-radius: 50% 20% / 10% 40%\">Human Amyotrophic Lateral Sclerosis (ALS) PBMCs</h1>\n\nPeripheral Blood Mononuclear Cells (PBMCs) are a subset of leukocytes (or White Blood Cell, WBC), which are integral components of the body’s immune system that fight disease. PBMCs consist of a mixture of lymphocytes (T cells, B cells, and NK cells) and monocytes and are isolated from peripheral whole blood via density gradient centrifugation and red blood cell lysis."},{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true,"_kg_hide-output":true,"_kg_hide-input":true,"collapsed":true},"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\nimport matplotlib\nimport matplotlib.pyplot as plt\nimport seaborn as sns\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\n\n# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# **<span style=\"color:#DC143C;\">Decreased Mitochondrial Function, Biogenesis, and Degradation in Peripheral Blood Mononuclear Cells from Amyotrophic Lateral Sclerosis Patients as a Potential Tool for Biomarker Research</span>**\n\nCitation: Araujo, B.G., Souza e Silva, L.F., de Barros Torresi, J.L. et al. Decreased Mitochondrial Function, Biogenesis, and Degradation in Peripheral Blood Mononuclear Cells from Amyotrophic Lateral Sclerosis Patients as a Potential Tool for Biomarker Research. Mol Neurobiol 57, 5084–5102 (2020). https://doi.org/10.1007/s12035-020-02059-1\n\n\"In PBMCs from ALS patients, the authors observed an overall lower mitochondrial calcium uptake/retention, mitochondrial depolarization, and redox homeostasis unbalance. Moreover, ALS PBMCs showed a significant decrease in critical metabolic genes and a diminishment in mitochondrial biogenesis and content. These findings seem to occur in parallel to an increase in mitochondrial fission and autophagy-related gene expression, despite a reduction in mitochondrial degradation signaling. Accordingly, ALS cells present a decrease in energy-producing metabolic compounds, such as ATP production and pyruvate, corroborating the fact that mitochondrial deregulation can lead to an energetic deficiency in peripheral cells as an ultimate consequence of the disorder. Thus, mitochondrial function evaluation in PBMCs could be a valuable strategy to detect ALS, as well as to assess its progression and therapy efficacy.\"\n\n\n**<span style=\"color:#DC143C;\">Depolarized Mitochondria and Disturbances in Calcium Handling in ALS PBMCs</span>**\n\n\"Considering the importance of mitochondria function in keeping calcium handling, and because modifications in the transport of electron and the proton motive force can interfere with ATP production, we evaluated the cytosolic Ca2+ level in PBMCs from ALS and control subjects. As the authors can notice, there is a significant reduction in the Fluo-4 fluorescence in the ALS group, suggesting a decrease in cytosolic calcium in the patient’s cells. Curiously, in the presence of FCCP (5 μM), there is also a diminishment in the cytosolic calcium in ALS PBMCs.\"\n\n\n**<span style=\"color:#DC143C;\">Decreased Antioxidant Defense and Metabolic Gene Expression in ALS PBMCs</span>**\n\n\"Because oxidative stress is increased in peripheral cells from ALS subjects, they further investigated the expression of NFE2L2, a transcription factor related to the synthesis of antioxidant enzymes. As shown in, there is a massive reduction in NFE2L2 in ALS cells, indicating that ALS PBMCs have a decreased antioxidant defense.\"\n\nhttps://link.springer.com/article/10.1007/s12035-020-02059-1"},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"df = pd.read_csv('../input/end-als/end-als/clinical-data/filtered-metadata/metadata/clinical/PBMC_Sample_Collection.csv', encoding='ISO-8859-2')\npd.set_option('display.max_columns', None)\ndf.head()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**<span style=\"color:#DC143C;\">Columns Names:</span>**\n\nsmplcol  = Collection number\n\nsmpldn = Status\n\nsmpldn = Collected 1, Not Collected 2\n\nsmpldnsp = Reason not collected\n\nsmpldt = Date collected\n\ntbscol = Number of tubes collected\n\ntubshdt = Date tubes shipped"},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"df.shape","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**<span style=\"color:#DC143C;\">Changes in Bioenergetics Metabolic Compounds in PMBCs from ALS Patients</span>**\n\n\"As the authors observed changes in mitochondrial function in ALS’ PBMCs, besides deviations in the expression of genes related to mitochondrial metabolism, biogenesis, and fission, they investigated whether the bioproducts related to energy metabolisms, such as ATP, ADP, pyruvate, and lactate, could vary. As observed, there is a significant decrease in ATP levels in PBMCs from ALS patients compared with control individuals; no modification is observed regarding ADP levels. As they can notice, the pyruvate level significantly dropped in ALS PBMCs, while the level of lactate is significantly elevated. This outcome suggests that pyruvate may be converted into lactate to produce energy since ATP synthesis, once oxidative phosphorylation might be jeopardized because of mitochondrial deregulation.\"\n\n\"Due to the increased ROS (Reactive oxygen species) level in ALS PBMCs, they decided to investigate NFE2L2 expression, which is implicated in cellular detoxifying systems. They observed a large diminishment in NFE2L2 expression (more than 50%) in ALS cells. The increased levels of ROS (Reactive oxygen species) could have been related to a failure of NFE2L2 in translocating to the nucleus, reducing the transcription of secondary factors of the antioxidant cascade. Interestingly, in previous work with PBMCs, the expression of NFE2L2 had no changes in relation to controls, but ALS patients were 10 years older than in their study, which could indicate a time-dependent NFE2L2 expression. In agreement with their outcome, the expression of NFE2L2 was declined in humans with ALS and in the SOD1-G93A model. Of importance, NFE2L2 overexpression was associated with more prolonged survival and neuroprotection in the SOD1-G93A mouse model.\"\n\n\"Their results indicate that PBMCs from ALS patients present a significant mitochondrial dysfunction that could be used as a powerful tool in investigating ALS disclosure and improvement. Importantly, the usage of PBMCs in biomarker research was already validated, and our findings can further contribute to future findings concerning new pharmacological strategies against ALS.\"\n\nhttps://link.springer.com/article/10.1007/s12035-020-02059-1"},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"df.isnull().sum()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**<span style=\"color:#DC143C;\">Amyotrophic Lateral Sclerosis Multiprotein Biomarkers in Peripheral Blood Mononuclear Cells</span>**\n\nAuthors: Giovanni Nardo, Silvia Pozzi, Mauro Pignataro, Eliana Lauranzano, Giorgia Spano, Silvia Garbelli, Stefania Mantovani, Kalliopi Marinou, Laura Papetti, Marta Monteforte, Valter Torri,  Gianfranco Bazzoni, Christian Lunetta, Massimo Corbo, Gabriele Mora, Caterina Bendotti, Valentina Bonetto.\nOctober 2011PLoS ONE 6(10):e25545 - DOI: 10.1371/journal.pone.0025545 - PubMed\n\n\" Specific molecular biomarkers are urgently needed to facilitate clinical studies and speed up the development of effective treatments. They used a two-dimensional difference in gel electrophoresis approach to identify in easily accessible clinical samples, peripheral blood mononuclear cells (PBMC), a panel of protein biomarkers that are closely associated with ALS. Validations and a longitudinal study were performed by immunoassays on a selected number of proteins. The same proteins were also measured in PBMC and spinal cord of a G93A SOD1 transgenic rat model.\" \n\n\"The authors identified combinations of protein biomarkers that can distinguish, with high discriminatory power, ALS patients from healthy controls (98%), and from patients with neurological disorders that may resemble ALS (91%), between two levels of disease severity (90%), and a number of translational biomarkers, that link responses between human and animal model. They demonstrated that TDP-43, cyclophilin A and ERp57 associate with disease progression in a longitudinal study. Moreover, the protein profile changes detected in peripheral blood mononuclear cells of ALS patients are suggestive of possible intracellular pathogenic mechanisms such as endoplasmic reticulum stress, nitrative stress, disturbances in redox regulation and RNA processing. Their results indicate that PBMC multiprotein biomarkers could contribute to determine amyotrophic lateral sclerosis diagnosis, differential diagnosis, disease severity and progression, and may help to elucidate pathogenic mechanisms.\"\n\nhttps://www.researchgate.net/publication/51716609_Amyotrophic_Lateral_Sclerosis_Multiprotein_Biomarkers_in_Peripheral_Blood_Mononuclear_Cells"},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"df[\"smpldnsp\"].value_counts()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"<h1 style=\"background-color:#DC143C; font-family:'Brush Script MT',cursive;color:white;font-size:200%; text-align:center;border-radius: 50% 20% / 10% 40%\">RNA-Seq profiling in peripheral blood mononuclear cells of amyotrophic lateral sclerosis patients and controls</h1>\n\nCitation: Zucca, S., Gagliardi, S., Pandini, C. et al. RNA-Seq profiling in peripheral blood mononuclear cells of amyotrophic lateral sclerosis patients and controls. Sci Data 6, 190006 (2019). https://doi.org/10.1038/sdata.2019.6\n\n\"Coding and long non-coding RNA (lncRNA) metabolism is now revealing its crucial role in Amyotrophic Lateral Sclerosis (ALS) pathogenesis. In that work, the authors present a dataset obtained via Illumina RNA-seq analysis on Peripheral Blood Mononuclear Cells (PBMCs) from sporadic and mutated ALS patients (mutations in FUS, TARDBP, SOD1 and VCP genes) and healthy controls. That dataset allows the whole-transcriptome characterization of PBMCs content, both in terms of coding and non-coding RNAs, in order to compare the disease state to the healthy controls, both for sporadic patients and for mutated patients. Their dataset is a starting point for the omni-comprehensive analysis of coding and lncRNAs, from an easy to withdraw, manage and store tissue that shows to be a suitable model for RNA profiling in ALS.\"\n\nhttps://www.nature.com/articles/sdata20196"},{"metadata":{},"cell_type":"markdown","source":"(a) The flowchart represents RNA-Seq workflow and data analysis. (b) An estimate of the dispersion parameter for each gene is shown. (c) Principal component analysis results. (d) Heatmap showing the sample-to-sample distance. It was obtained with DeSeq2 package on regularized-logarithm transformed counts. Color code is reported above the heatmap.\n\nCitation: Zucca, S., Gagliardi, S., Pandini, C. et al. RNA-Seq profiling in peripheral blood mononuclear cells of amyotrophic lateral sclerosis patients and controls. Sci Data 6, 190006 (2019). https://doi.org/10.1038/sdata.2019.6\n\n![](https://media.springernature.com/lw685/springer-static/image/art%3A10.1038%2Fsdata.2019.6/MediaObjects/41597_2019_Article_BFsdata20196_Fig2_HTML.jpg?as=webp)https://www.nature.com/articles/sdata20196"},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"#Tubes Collected 1, Not Collected 2\n#Codes by https://www.kaggle.com/muhammadismail99/data-analysis-of-guf-pub-dataset/notebook\n\nsns.countplot(data = df, x = 'smpldn')\nplt.title('Tubes collected')","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-output":true,"trusted":true,"collapsed":true},"cell_type":"code","source":"#https://auto.gluon.ai/stable/tutorials/tabular_prediction/tabular-quickstart.html\n\n#Code by Daiki Katsuragawa https://www.kaggle.com/daikikatsuragawa/tps-apr-2021-benchmark-using-autogluon/notebook\n\n!pip install autogluon.tabular[all]","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"from autogluon.tabular import TabularDataset, TabularPredictor","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-output":true,"trusted":true,"collapsed":true},"cell_type":"code","source":"#https://auto.gluon.ai/stable/tutorials/tabular_prediction/tabular-quickstart.html\n\n#Code by Daiki Katsuragawa https://www.kaggle.com/daikikatsuragawa/tps-apr-2021-benchmark-using-autogluon/notebook\n\nlabel = 'smpldn'\ntime_limit = 3600\npredictor = TabularPredictor(label=label).fit(df, time_limit=time_limit, num_bag_folds=8)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#For tabular problems, fit() returns a Predictor object. For classification, you can easily output predicted class probabilities instead of predicted classes:\nhttps://auto.gluon.ai/stable/tutorials/tabular_prediction/tabular-quickstart.html"},{"metadata":{"trusted":true},"cell_type":"code","source":"#https://auto.gluon.ai/stable/tutorials/tabular_prediction/tabular-quickstart.html\n\npred_probs = predictor.predict_proba(df)\npred_probs.head(5)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"results = predictor.fit_summary()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"From this summary, we can see that AutoGluon trained many different types of models as well as an ensemble of the best-performing models. The summary also describes the actual models that were trained during fit and how well each model performed on the held-out validation data. We can view what properties AutoGluon automatically inferred about our prediction task:"},{"metadata":{"trusted":true},"cell_type":"code","source":"print(\"AutoGluon infers problem type is: \", predictor.problem_type)\nprint(\"AutoGluon identified the following types of features:\")\nprint(predictor.feature_metadata)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#Evaluate the performance of each individual trained model on our (labeled) data."},{"metadata":{"trusted":true},"cell_type":"code","source":"predictor.leaderboard(df, silent=True)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"When we call predict(), AutoGluon automatically predicts with the model that displayed the best performance on validation data (i.e. the weighted-ensemble). We can instead specify which model to use for predictions like this:"},{"metadata":{"trusted":true},"cell_type":"code","source":"predictor.predict(df, model='LightGBM_BAG_L1')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"You should not call fit() with entirely default arguments if you are benchmarking AutoGluon-Tabular or hoping to maximize its accuracy! To get the best predictive accuracy with AutoGluon, you should generally use it like this:"},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"time_limit = 60  # for quick demonstration only, you should set this to longest time you are willing to wait (in seconds)\nmetric = 'roc_auc'  # specify your evaluation metric here\npredictor = TabularPredictor(label, eval_metric=metric).fit(df, time_limit=time_limit, presets='best_quality')\npredictor.leaderboard(df, silent=True)","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}