{"cells":[{"metadata":{},"cell_type":"markdown","source":"### problem statement - find closest answers to key questions from a corpus of scientific papers ###\n\ngoal:\ngiven the urgency of helping find a cure within the many scientific papers released, the goal here is to extract relevant answers to the questions posed so that the scientific community can pinpoint its research.\n\n<img src=\"https://raw.githubusercontent.com/gisblog/nih-covid19/master/covid19/kaggle/working/answers.task.3.biorxiv_medrxiv.json.jpg\" />\n#### selected places in the extracted NLP answers for task 3 - \"HELP US UNDERSTAND HOW GEOGRAPHY AFFECTS VIRALITY?\" ####\n\nsteps:\n* collect tasks' QUESTIONS.\n* walk given path for papers.\n* if conditions are met, then write paper paths to a main papers file.\n* given a list of papers in the main file, write the top unique answers to a task's QUESTIONS by evaluating the papers:\n- for each paper, do -\n- collect each QUESTION and relevant elements from the paper.\n- convert the collection of words to vectors.\n- calculate euclidean distances from the QUESTIONS to the paper elements, and sort by similarity/distance.\n- from the sorting, remove the original QUESTIONS and write the top unique answers to an answer file.\n* walk given path for answer files.\n* if conditions are met, then merge all answers on a given path by task # into a main answer file.\n>     question 1\n>     question 2\n>     question 3\n>         +\n>         <-------> question +-> vectors\n>         +         paper        (10101) +\n>     paper 1                            |\n>     paper 2                            v\n>     paper 3                    sort answers by similarity\n>                                get unique answers\n>                                        +\n>                                        |\n>                                        v\n>                                    +---+----+\n>                                    | answer |\n>                                    +--------+\n>                                     +\n>                                     |_+\n>                                       |_+\n>                                       |\n>                                     +-v--+\n>                                     |main|\n>                                     +----+\n(the dhyana of python - tim peters)\n\ndataset used:\nCOVID-19 Open Research Dataset (CORD-19) initially released by the white house and its coalition of leading research groups that comprised of 13,202 scientific papers broken down into 4 subsets by source type - biorxiv_medrxiv, comm_use_subset, noncomm_use_subset and pmc_custom_license.\n\nnotebook trimmed to run on kaggle. see full code on github below."},{"metadata":{},"cell_type":"markdown","source":"**from packages, import required modules ↓**\n\n**NLP: pros and cons of using CountVectorizer over HashingVectorizer (or TfidfVectorizer)**\n* CountVectorizer uses in-memory vocabulary.\n* HashingVectorizer doesn't have a way to compute the inverse transform (from feature indices to string feature names).\n* This can be a problem when trying to introspect which features are most important to a model.\n* Also, no idf (inverse document frequency) weighting -\n* idf measures how important a word is to a doc in a collection of docs.\n* tf–idf increases with the # of times a word appears in a doc.\n* tf-idf decreases with the # of docs in the collection that contain the word."},{"metadata":{"_uuid":"d629ff2d2480ee46fbb7e2d37f6b5fab8052498a","_cell_guid":"79c7e3d0-c299-4dcb-8224-4455121ee9b0","trusted":true,"_kg_hide-output":true,"scrolled":true},"cell_type":"code","source":"from sklearn.feature_extraction.text import CountVectorizer # convert collection of txt docs to matrix of token counts\nfrom sklearn.metrics.pairwise import euclidean_distances # compute distance matrix between each pair of vectors\nfrom datetime import datetime # timestamps\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, csv file i/o (e.g. pd.read_csv)\nimport os # misc os interfaces\nimport json # json encoder + decoder\n# import jsonstreams # writes json as stream\nimport sys # for constants, functions and methods of py interpreter\nimport re # regex\nimport pprint # pretty print\n# import jsbeautifier # beautify, unpack or deobfuscate js\n# import threading # i/o operations\nimport concurrent.futures # asynchronously execute callables with threads or processes\nimport multiprocessing # cpu-heavy operations\nimport psutil # process and system monitoring\n# import logging # event logging for apps and libs","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**collect tasks' QUESTIONS ↓**\n\nfor each task-question at https://www.kaggle.com/allen-institute-for-ai/CORD-19-research-challenge/tasks, create a list.\nalso create a detailed-question list and a specific-question list.\nnote, QUESTIONS are capitalized to better visually spot them from the answers during testing."},{"metadata":{"trusted":true,"scrolled":true},"cell_type":"code","source":"### tasks/questions ###\nquestions = [\n    'WHAT IS KNOWN ABOUT TRANSMISSION, INCUBATION, AND ENVIRONMENTAL STABILITY?',\n    'WHAT DO WE KNOW ABOUT COVID-19 RISK FACTORS?',\n    'WHAT DO WE KNOW ABOUT VIRUS GENETICS, ORIGIN, AND EVOLUTION?',\n    'HELP US UNDERSTAND HOW GEOGRAPHY AFFECTS VIRALITY?',\n    'WHAT DO WE KNOW ABOUT VACCINES AND THERAPEUTICS?',\n    'WHAT DO WE KNOW ABOUT NON-PHARMACEUTICAL INTERVENTIONS?',\n    'WHAT HAS BEEN PUBLISHED ABOUT MEDICAL CARE?',\n    'WHAT HAS BEEN PUBLISHED ABOUT ETHICAL AND SOCIAL SCIENCE CONSIDERATIONS?',\n    'WHAT HAS BEEN PUBLISHED ABOUT INFORMATION SHARING AND INTER-SECTORAL COLLABORATION?',\n    'WHAT DO WE KNOW ABOUT DIAGNOSTICS AND SURVEILLANCE?'\n    ]\n\n### detailed questions ###\n# answers_0_detail = ['X is known about transmission, incubation, and environmental stability.'] # test\nquestions_0_detail = [\n    'WHAT DO WE KNOW ABOUT NATURAL HISTORY, TRANSMISSION, AND DIAGNOSTICS FOR THE VIRUS?',\n    'WHAT HAVE WE LEARNED ABOUT INFECTION PREVENTION AND CONTROL?'\n    ]\nquestions_1_detail = [\n    'WHAT HAVE WE LEARNED FROM EPIDEMIOLOGICAL STUDIES?'\n    ]\nquestions_2_detail = [\n    'WHAT DO WE KNOW ABOUT THE VIRUS ORIGIN AND MANAGEMENT MEASURES AT THE HUMAN-ANIMAL INTERFACE?'\n]\nquestions_3_detail = [\n    'ARE THERE GEOGRAPHIC VARIATIONS IN HOW THE DISEASE WILL SPREAD?',\n    'ARE THERE DIFFERENT VARIATIONS OF THE VIRUS IN DIFFERENT AREAS?'\n]\nquestions_4_detail = [\n    'WHAT HAS BEEN PUBLISHED CONCERNING RESEARCH AND DEVELOPMENT AND EVALUATION EFFORTS OF VACCINES AND THERAPEUTICS?'\n]\nquestions_5_detail = [\n    'WHAT DO WE KNOW ABOUT THE EFFECTIVENESS OF NON-PHARMACEUTICAL INTERVENTIONS?',\n    'WHAT IS KNOWN ABOUT EQUITY AND BARRIERS TO COMPLIANCE FOR NON-PHARMACEUTICAL INTERVENTIONS?'\n]\nquestions_6_detail = [\n    'WHAT HAS BEEN PUBLISHED CONCERNING SURGE CAPACITY AND NURSING HOMES?',\n    'WHAT HAS BEEN PUBLISHED CONCERNING EFFORTS TO INFORM ALLOCATION OF SCARCE RESOURCES?',\n    'WHAT DO WE KNOW ABOUT PERSONAL PROTECTIVE EQUIPMENT?',\n    'WHAT HAS BEEN PUBLISHED CONCERNING ALTERNATIVE METHODS TO ADVISE ON DISEASE MANAGEMENT?',\n    'WHAT HAS BEEN PUBLISHED CONCERNING PROCESSES OF CARE?',\n    'WHAT DO WE KNOW ABOUT THE CLINICAL CHARACTERIZATION AND MANAGEMENT OF THE VIRUS?'\n]\nquestions_7_detail = [\n    'WHAT HAS BEEN PUBLISHED CONCERNING ETHICAL CONSIDERATIONS FOR RESEARCH?',\n    'WHAT HAS BEEN PUBLISHED CONCERNING SOCIAL SCIENCES AT THE OUTBREAK RESPONSE?'\n]\nquestions_8_detail = [\n    'WHAT HAS BEEN PUBLISHED ABOUT DATA STANDARDS AND NOMENCLATURE?',\n    'WHAT HAS BEEN PUBLISHED ABOUT GOVERNMENTAL PUBLIC HEALTH? WHAT DO WE KNOW ABOUT RISK COMMUNICATION?',\n    'WHAT HAS BEEN PUBLISHED ABOUT COMMUNICATING WITH HIGH-RISK POPULATIONS?',\n    'WHAT HAS BEEN PUBLISHED TO CLARIFY COMMUNITY MEASURES?',\n    'WHAT HAS BEEN PUBLISHED ABOUT EQUITY CONSIDERATIONS AND PROBLEMS OF INEQUITY?'\n]\nquestions_9_detail = [\n    'WHAT HAS BEEN PUBLISHED CONCERNING SYSTEMATIC, HOLISTIC APPROACH TO DIAGNOSTICS (FROM THE PUBLIC HEALTH SURVEILLANCE PERSPECTIVE TO BEING ABLE TO PREDICT CLINICAL OUTCOMES)?'\n]\n###\nquestions_detail = [questions_0_detail, questions_1_detail, questions_2_detail, questions_3_detail, questions_4_detail, questions_5_detail, questions_6_detail, questions_7_detail, questions_8_detail, questions_9_detail]\n\n### specific questions ###\nquestions_0_specific = [\n    'RANGE OF INCUBATION PERIODS FOR THE DISEASE IN HUMANS (AND HOW THIS VARIES ACROSS AGE AND HEALTH STATUS) AND HOW LONG INDIVIDUALS ARE CONTAGIOUS, EVEN AFTER RECOVERY?',\n    'PREVALENCE OF ASYMPTOMATIC SHEDDING AND TRANSMISSION (E.G., PARTICULARLY CHILDREN)?',\n    'SEASONALITY OF TRANSMISSION?',\n    'PHYSICAL SCIENCE OF THE CORONAVIRUS (E.G., CHARGE DISTRIBUTION, ADHESION TO HYDROPHILIC/PHOBIC SURFACES, ENVIRONMENTAL SURVIVAL TO INFORM DECONTAMINATION EFFORTS FOR AFFECTED AREAS AND PROVIDE INFORMATION ABOUT VIRAL SHEDDING)?',\n    'PERSISTENCE AND STABILITY ON A MULTITUDE OF SUBSTRATES AND SOURCES (E.G., NASAL DISCHARGE, SPUTUM, URINE, FECAL MATTER, BLOOD)?',\n    'PERSISTENCE OF VIRUS ON SURFACES OF DIFFERENT MATERIALS (E,G., COPPER, STAINLESS STEEL, PLASTIC)?',\n    'NATURAL HISTORY OF THE VIRUS AND SHEDDING OF IT FROM AN INFECTED PERSON?',\n    'IMPLEMENTATION OF DIAGNOSTICS AND PRODUCTS TO IMPROVE CLINICAL PROCESSES?'\n    'DISEASE MODELS, INCLUDING ANIMAL MODELS FOR INFECTION, DISEASE AND TRANSMISSION?',\n    'TOOLS AND STUDIES TO MONITOR PHENOTYPIC CHANGE AND POTENTIAL ADAPTATION OF THE VIRUS?',\n    'IMMUNE RESPONSE AND IMMUNITY?',\n    'EFFECTIVENESS OF MOVEMENT CONTROL STRATEGIES TO PREVENT SECONDARY TRANSMISSION IN HEALTH CARE AND COMMUNITY SETTINGS?',\n    'EFFECTIVENESS OF PERSONAL PROTECTIVE EQUIPMENT (PPE) AND ITS USEFULNESS TO REDUCE RISK OF TRANSMISSION IN HEALTH CARE AND COMMUNITY SETTINGS?',\n    'ROLE OF THE ENVIRONMENT IN TRANSMISSION?'\n    ]\nquestions_1_specific = [\n    'DATA ON POTENTIAL RISKS FACTORS?',\n    'SMOKING, PRE-EXISTING PULMONARY DISEASE?',\n    'CO-INFECTIONS (DETERMINE WHETHER CO-EXISTING RESPIRATORY/VIRAL INFECTIONS MAKE THE VIRUS MORE TRANSMISSIBLE OR VIRULENT) AND OTHER CO-MORBIDITIES?',\n    'NEONATES AND PREGNANT WOMEN?',\n    'SOCIO-ECONOMIC AND BEHAVIORAL FACTORS TO UNDERSTAND THE ECONOMIC IMPACT OF THE VIRUS AND WHETHER THERE WERE DIFFERENCES?',\n    'TRANSMISSION DYNAMICS OF THE VIRUS, INCLUDING THE BASIC REPRODUCTIVE NUMBER, INCUBATION PERIOD, SERIAL INTERVAL, MODES OF TRANSMISSION AND ENVIRONMENTAL FACTORS?',\n    'SEVERITY OF DISEASE, INCLUDING RISK OF FATALITY AMONG SYMPTOMATIC HOSPITALIZED PATIENTS, AND HIGH-RISK PATIENT GROUPS?',\n    'SUSCEPTIBILITY OF POPULATIONS?',\n    'PUBLIC HEALTH MITIGATION MEASURES THAT COULD BE EFFECTIVE FOR CONTROL?'\n]\nquestions_2_specific = [\n    'REAL-TIME TRACKING OF WHOLE GENOMES AND A MECHANISM FOR COORDINATING THE RAPID DISSEMINATION OF THAT INFORMATION TO INFORM THE DEVELOPMENT OF DIAGNOSTICS AND THERAPEUTICS AND TO TRACK VARIATIONS OF THE VIRUS OVER TIME?',\n    'ACCESS TO GEOGRAPHIC AND TEMPORAL DIVERSE SAMPLE SETS TO UNDERSTAND GEOGRAPHIC DISTRIBUTION AND GENOMIC DIFFERENCES, AND DETERMINE WHETHER THERE IS MORE THAN ONE STRAIN IN CIRCULATION. MULTI-LATERAL AGREEMENTS SUCH AS THE NAGOYA PROTOCOL COULD BE LEVERAGED?',\n    'EVIDENCE THAT LIVESTOCK COULD BE INFECTED (E.G., FIELD SURVEILLANCE, GENETIC SEQUENCING, RECEPTOR BINDING) AND SERVE AS A RESERVOIR AFTER THE EPIDEMIC APPEARS TO BE OVER?',\n    'EVIDENCE OF WHETHER FARMERS ARE INFECTED, AND WHETHER FARMERS COULD HAVE PLAYED A ROLE IN THE ORIGIN?',\n    'SURVEILLANCE OF MIXED WILDLIFE- LIVESTOCK FARMS FOR SARS-COV-2 AND OTHER CORONAVIRUSES IN SOUTHEAST ASIA?',\n    'EXPERIMENTAL INFECTIONS TO TEST HOST RANGE FOR THIS PATHOGEN?',\n    'ANIMAL HOST(S) AND ANY EVIDENCE OF CONTINUED SPILL-OVER TO HUMANS?',\n    'SOCIOECONOMIC AND BEHAVIORAL RISK FACTORS FOR THIS SPILL-OVER?',\n    'SUSTAINABLE RISK REDUCTION STRATEGIES?'\n]\nquestions_3_specific = [\n    'ARE THERE GEOGRAPHIC VARIATIONS IN THE RATE OF COVID-19 SPREAD?',\n    'ARE THERE GEOGRAPHIC VARIATIONS IN THE MORTALITY RATE OF COVID-19?',\n    'IS THERE ANY EVIDENCE TO SUGGEST GEOGRAPHIC BASED VIRUS MUTATIONS?'\n]\nquestions_4_specific = [\n    'EFFECTIVENESS OF DRUGS BEING DEVELOPED AND TRIED TO TREAT COVID-19 PATIENTS?',\n    'CLINICAL AND BENCH TRIALS TO INVESTIGATE LESS COMMON VIRAL INHIBITORS AGAINST COVID-19 SUCH AS NAPROXEN, CLARITHROMYCIN, AND MINOCYCLINETHAT THAT MAY EXERT EFFECTS ON VIRAL REPLICATION?',\n    'METHODS EVALUATING POTENTIAL COMPLICATION OF ANTIBODY-DEPENDENT ENHANCEMENT (ADE) IN VACCINE RECIPIENTS?',\n    'EXPLORATION OF USE OF BEST ANIMAL MODELS AND THEIR PREDICTIVE VALUE FOR A HUMAN VACCINE?',\n    'CAPABILITIES TO DISCOVER A THERAPEUTIC (NOT VACCINE) FOR THE DISEASE, AND CLINICAL EFFECTIVENESS STUDIES TO DISCOVER THERAPEUTICS, TO INCLUDE ANTIVIRAL AGENTS?',\n    'ALTERNATIVE MODELS TO AID DECISION MAKERS IN DETERMINING HOW TO PRIORITIZE AND DISTRIBUTE SCARCE, NEWLY PROVEN THERAPEUTICS AS PRODUCTION RAMPS UP. THIS COULD INCLUDE IDENTIFYING APPROACHES FOR EXPANDING PRODUCTION CAPACITY TO ENSURE EQUITABLE AND TIMELY DISTRIBUTION TO POPULATIONS IN NEED?',\n    'EFFORTS TARGETED AT A UNIVERSAL CORONAVIRUS VACCINE?',\n    'EFFORTS TO DEVELOP ANIMAL MODELS AND STANDARDIZE CHALLENGE STUDIES?',\n    'EFFORTS TO DEVELOP PROPHYLAXIS CLINICAL STUDIES AND PRIORITIZE IN HEALTHCARE WORKERS?',\n    'APPROACHES TO EVALUATE RISK FOR ENHANCED DISEASE AFTER VACCINATION?',\n    'ASSAYS TO EVALUATE VACCINE IMMUNE RESPONSE AND PROCESS DEVELOPMENT FOR VACCINES, ALONGSIDE SUITABLE ANIMAL MODELS [IN CONJUNCTION WITH THERAPEUTICS]?'\n]\nquestions_5_specific = [\n    'GUIDANCE ON WAYS TO SCALE UP NPIS IN A MORE COORDINATED WAY (E.G., ESTABLISH FUNDING, INFRASTRUCTURE AND AUTHORITIES TO SUPPORT REAL TIME, AUTHORITATIVE (QUALIFIED PARTICIPANTS) COLLABORATION WITH ALL STATES TO GAIN CONSENSUS ON CONSISTENT GUIDANCE AND TO MOBILIZE RESOURCES TO GEOGRAPHIC AREAS WHERE CRITICAL SHORTFALLS ARE IDENTIFIED) TO GIVE US TIME TO ENHANCE OUR HEALTH CARE DELIVERY SYSTEM CAPACITY TO RESPOND TO AN INCREASE IN CASES?',\n    'RAPID DESIGN AND EXECUTION OF EXPERIMENTS TO EXAMINE AND COMPARE NPIS CURRENTLY BEING IMPLEMENTED. DHS CENTERS FOR EXCELLENCE COULD POTENTIALLY BE LEVERAGED TO CONDUCT THESE EXPERIMENTS?',\n    'RAPID ASSESSMENT OF THE LIKELY EFFICACY OF SCHOOL CLOSURES, TRAVEL BANS, BANS ON MASS GATHERINGS OF VARIOUS SIZES, AND OTHER SOCIAL DISTANCING APPROACHES?',\n    'METHODS TO CONTROL THE SPREAD IN COMMUNITIES, BARRIERS TO COMPLIANCE AND HOW THESE VARY AMONG DIFFERENT POPULATIONS?',\n    'MODELS OF POTENTIAL INTERVENTIONS TO PREDICT COSTS AND BENEFITS THAT TAKE ACCOUNT OF SUCH FACTORS AS RACE, INCOME, DISABILITY, AGE, GEOGRAPHIC LOCATION, IMMIGRATION STATUS, HOUSING STATUS, EMPLOYMENT STATUS, AND HEALTH INSURANCE STATUS?',\n    'POLICY CHANGES NECESSARY TO ENABLE THE COMPLIANCE OF INDIVIDUALS WITH LIMITED RESOURCES AND THE UNDERSERVED WITH NPIS?',\n    'RESEARCH ON WHY PEOPLE FAIL TO COMPLY WITH PUBLIC HEALTH ADVICE, EVEN IF THEY WANT TO DO SO (E.G., SOCIAL OR FINANCIAL COSTS MAY BE TOO HIGH)?',\n    'RESEARCH ON THE ECONOMIC IMPACT OF THIS OR ANY PANDEMIC. THIS WOULD INCLUDE IDENTIFYING POLICY AND PROGRAMMATIC ALTERNATIVES THAT LESSEN/MITIGATE RISKS TO CRITICAL GOVERNMENT SERVICES, FOOD DISTRIBUTION AND SUPPLIES, ACCESS TO CRITICAL HOUSEHOLD SUPPLIES, AND ACCESS TO HEALTH DIAGNOSES, TREATMENT, AND NEEDED CARE, REGARDLESS OF ABILITY TO PAY?'\n]\nquestions_6_specific = [\n    'RESOURCES TO SUPPORT SKILLED NURSING FACILITIES AND LONG TERM CARE FACILITIES?',\n    'MOBILIZATION OF SURGE MEDICAL STAFF TO ADDRESS SHORTAGES IN OVERWHELMED COMMUNITIES?',\n    'AGE-ADJUSTED MORTALITY DATA FOR ACUTE RESPIRATORY DISTRESS SYNDROME (ARDS) WITH/WITHOUT OTHER ORGAN FAILURE – PARTICULARLY FOR VIRAL ETIOLOGIES?',\n    'EXTRACORPOREAL MEMBRANE OXYGENATION (ECMO) OUTCOMES DATA OF COVID-19 PATIENTS?',\n    'OUTCOMES DATA FOR COVID-19 AFTER MECHANICAL VENTILATION ADJUSTED FOR AGE?',\n    'KNOWLEDGE OF THE FREQUENCY, MANIFESTATIONS, AND COURSE OF EXTRAPULMONARY MANIFESTATIONS OF COVID-19, INCLUDING, BUT NOT LIMITED TO, POSSIBLE CARDIOMYOPATHY AND CARDIAC ARREST?',\n    'APPLICATION OF REGULATORY STANDARDS (E.G., EUA, CLIA) AND ABILITY TO ADAPT CARE TO CRISIS STANDARDS OF CARE LEVEL?',\n    'APPROACHES FOR ENCOURAGING AND FACILITATING THE PRODUCTION OF ELASTOMERIC RESPIRATORS, WHICH CAN SAVE THOUSANDS OF N95 MASKS?',\n    'BEST TELEMEDICINE PRACTICES, BARRIERS AND FACIITATORS, AND SPECIFIC ACTIONS TO REMOVE/EXPAND THEM WITHIN AND ACROSS STATE BOUNDARIES?',\n    'GUIDANCE ON THE SIMPLE THINGS PEOPLE CAN DO AT HOME TO TAKE CARE OF SICK PEOPLE AND MANAGE DISEASE?',\n    'ORAL MEDICATIONS THAT MIGHT POTENTIALLY WORK?',\n    'USE OF AI IN REAL-TIME HEALTH CARE DELIVERY TO EVALUATE INTERVENTIONS, RISK FACTORS, AND OUTCOMES IN A WAY THAT COULD NOT BE DONE MANUALLY?',\n    'BEST PRACTICES AND CRITICAL CHALLENGES AND INNOVATIVE SOLUTIONS AND TECHNOLOGIES IN HOSPITAL FLOW AND ORGANIZATION, WORKFORCE PROTECTION, WORKFORCE ALLOCATION, COMMUNITY-BASED SUPPORT RESOURCES, PAYMENT, AND SUPPLY CHAIN MANAGEMENT TO ENHANCE CAPACITY, EFFICIENCY, AND OUTCOMES?',\n    'EFFORTS TO DEFINE THE NATURAL HISTORY OF DISEASE TO INFORM CLINICAL CARE, PUBLIC HEALTH INTERVENTIONS, INFECTION PREVENTION CONTROL, TRANSMISSION, AND CLINICAL TRIALS?',\n    'EFFORTS TO DEVELOP A CORE CLINICAL OUTCOME SET TO MAXIMIZE USABILITY OF DATA ACROSS A RANGE OF TRIALS?',\n    'EFFORTS TO DETERMINE ADJUNCTIVE AND SUPPORTIVE INTERVENTIONS THAT CAN IMPROVE THE CLINICAL OUTCOMES OF INFECTED PATIENTS (E.G. STEROIDS, HIGH FLOW OXYGEN)?'\n]\nquestions_7_specific = [\n    'EFFORTS TO ARTICULATE AND TRANSLATE EXISTING ETHICAL PRINCIPLES AND STANDARDS TO SALIENT ISSUES IN COVID-2019?',\n    'EFFORTS TO EMBED ETHICS ACROSS ALL THEMATIC AREAS, ENGAGE WITH NOVEL ETHICAL ISSUES THAT ARISE AND COORDINATE TO MINIMIZE DUPLICATION OF OVERSIGHT?',\n    'EFFORTS TO SUPPORT SUSTAINED EDUCATION, ACCESS, AND CAPACITY BUILDING IN THE AREA OF ETHICS?',\n    'EFFORTS TO ESTABLISH A TEAM AT WHO THAT WILL BE INTEGRATED WITHIN MULTIDISCIPLINARY RESEARCH AND OPERATIONAL PLATFORMS AND THAT WILL CONNECT WITH EXISTING AND EXPANDED GLOBAL NETWORKS OF SOCIAL SCIENCES?',\n    'EFFORTS TO DEVELOP QUALITATIVE ASSESSMENT FRAMEWORKS TO SYSTEMATICALLY COLLECT INFORMATION RELATED TO LOCAL BARRIERS AND ENABLERS FOR THE UPTAKE AND ADHERENCE TO PUBLIC HEALTH MEASURES FOR PREVENTION AND CONTROL. THIS INCLUDES THE RAPID IDENTIFICATION OF THE SECONDARY IMPACTS OF THESE MEASURES. (E.G. USE OF SURGICAL MASKS, MODIFICATION OF HEALTH SEEKING BEHAVIORS FOR SRH, SCHOOL CLOSURES)?',\n    'EFFORTS TO IDENTIFY HOW THE BURDEN OF RESPONDING TO THE OUTBREAK AND IMPLEMENTING PUBLIC HEALTH MEASURES AFFECTS THE PHYSICAL AND PSYCHOLOGICAL HEALTH OF THOSE PROVIDING CARE FOR COVID-19 PATIENTS AND IDENTIFY THE IMMEDIATE NEEDS THAT MUST BE ADDRESSED?',\n    'EFFORTS TO IDENTIFY THE UNDERLYING DRIVERS OF FEAR, ANXIETY AND STIGMA THAT FUEL MISINFORMATION AND RUMOR, PARTICULARLY THROUGH SOCIAL MEDIA?'\n]\nquestions_8_specific = [\n    'METHODS FOR COORDINATING DATA-GATHERING WITH STANDARDIZED NOMENCLATURE?',\n    'SHARING RESPONSE INFORMATION AMONG PLANNERS, PROVIDERS, AND OTHERS?',\n    'UNDERSTANDING AND MITIGATING BARRIERS TO INFORMATION-SHARING?',\n    'HOW TO RECRUIT, SUPPORT, AND COORDINATE LOCAL (NON-FEDERAL) EXPERTISE AND CAPACITY RELEVANT TO PUBLIC HEALTH EMERGENCY RESPONSE (PUBLIC, PRIVATE, COMMERCIAL AND NON-PROFIT, INCLUDING ACADEMIC)?',\n    'INTEGRATION OF FEDERAL/STATE/LOCAL PUBLIC HEALTH SURVEILLANCE SYSTEMS?',\n    'VALUE OF INVESTMENTS IN BASELINE PUBLIC HEALTH RESPONSE INFRASTRUCTURE PREPAREDNESS?',\n    'MODES OF COMMUNICATING WITH TARGET HIGH-RISK POPULATIONS (ELDERLY, HEALTH CARE WORKERS)?',\n    'RISK COMMUNICATION AND GUIDELINES THAT ARE EASY TO UNDERSTAND AND FOLLOW (INCLUDE TARGETING AT RISK POPULATIONS’ FAMILIES TOO)?',\n    'COMMUNICATION THAT INDICATES POTENTIAL RISK OF DISEASE TO ALL POPULATION GROUPS?',\n    'MISUNDERSTANDING AROUND CONTAINMENT AND MITIGATION?',\n    'ACTION PLAN TO MITIGATE GAPS AND PROBLEMS OF INEQUITY IN THE NATION’S PUBLIC HEALTH CAPABILITY, CAPACITY, AND FUNDING TO ENSURE ALL CITIZENS IN NEED ARE SUPPORTED AND CAN ACCESS INFORMATION, SURVEILLANCE, AND TREATMENT?',\n    'MEASURES TO REACH MARGINALIZED AND DISADVANTAGED POPULATIONS?',\n    'DATA SYSTEMS AND RESEARCH PRIORITIES AND AGENDAS INCORPORATE ATTENTION TO THE NEEDS AND CIRCUMSTANCES OF DISADVANTAGED POPULATIONS AND UNDERREPRESENTED MINORITIES?',\n    'MITIGATING THREATS TO INCARCERATED PEOPLE FROM COVID-19, ASSURING ACCESS TO INFORMATION, PREVENTION, DIAGNOSIS, AND TREATMENT?',\n    'UNDERSTANDING COVERAGE POLICIES (BARRIERS AND OPPORTUNITIES) RELATED TO TESTING, TREATMENT, AND CARE?'\n]\nquestions_9_specific = [\n    'HOW WIDESPREAD CURRENT EXPOSURE IS TO BE ABLE TO MAKE IMMEDIATE POLICY RECOMMENDATIONS ON MITIGATION MEASURES. DENOMINATORS FOR TESTING AND A MECHANISM FOR RAPIDLY SHARING THAT INFORMATION, INCLUDING DEMOGRAPHICS, TO THE EXTENT POSSIBLE. SAMPLING METHODS TO DETERMINE ASYMPTOMATIC DISEASE (E.G., USE OF SEROSURVEYS (SUCH AS CONVALESCENT SAMPLES) AND EARLY DETECTION OF DISEASE (E.G., USE OF SCREENING OF NEUTRALIZING ANTIBODIES SUCH AS ELISAS)?',\n    'EFFORTS TO INCREASE CAPACITY ON EXISTING DIAGNOSTIC PLATFORMS AND TAP INTO EXISTING SURVEILLANCE PLATFORMS?',\n    'RECRUITMENT, SUPPORT, AND COORDINATION OF LOCAL EXPERTISE AND CAPACITY (PUBLIC, PRIVATE—COMMERCIAL, AND NON-PROFIT, INCLUDING ACADEMIC), INCLUDING LEGAL, ETHICAL, COMMUNICATIONS, AND OPERATIONAL ISSUES?',\n    'NATIONAL GUIDANCE AND GUIDELINES ABOUT BEST PRACTICES TO STATES (E.G., HOW STATES MIGHT LEVERAGE UNIVERSITIES AND PRIVATE LABORATORIES FOR TESTING PURPOSES, COMMUNICATIONS TO PUBLIC HEALTH OFFICIALS AND THE PUBLIC)?',\n    'DEVELOPMENT OF A POINT-OF-CARE TEST (LIKE A RAPID INFLUENZA TEST) AND RAPID BED-SIDE TESTS, RECOGNIZING THE TRADEOFFS BETWEEN SPEED, ACCESSIBILITY, AND ACCURACY?',\n    'RAPID DESIGN AND EXECUTION OF TARGETED SURVEILLANCE EXPERIMENTS CALLING FOR ALL POTENTIAL TESTERS USING PCR IN A DEFINED AREA TO START TESTING AND REPORT TO A SPECIFIC ENTITY. THESE EXPERIMENTS COULD AID IN COLLECTING LONGITUDINAL SAMPLES, WHICH ARE CRITICAL TO UNDERSTANDING THE IMPACT OF AD HOC LOCAL INTERVENTIONS (WHICH ALSO NEED TO BE RECORDED)?',\n    'SEPARATION OF ASSAY DEVELOPMENT ISSUES FROM INSTRUMENTS, AND THE ROLE OF THE PRIVATE SECTOR TO HELP QUICKLY MIGRATE ASSAYS ONTO THOSE DEVICES?',\n    'EFFORTS TO TRACK THE EVOLUTION OF THE VIRUS (I.E., GENETIC DRIFT OR MUTATIONS) AND AVOID LOCKING INTO SPECIFIC REAGENTS AND SURVEILLANCE/DETECTION SCHEMES?',\n    'LATENCY ISSUES AND WHEN THERE IS SUFFICIENT VIRAL LOAD TO DETECT THE PATHOGEN, AND UNDERSTANDING OF WHAT IS NEEDED IN TERMS OF BIOLOGICAL AND ENVIRONMENTAL SAMPLING?',\n    'USE OF DIAGNOSTICS SUCH AS HOST RESPONSE MARKERS (E.G., CYTOKINES) TO DETECT EARLY DISEASE OR PREDICT SEVERE DISEASE PROGRESSION, WHICH WOULD BE IMPORTANT TO UNDERSTANDING BEST CLINICAL PRACTICE AND EFFICACY OF THERAPEUTIC INTERVENTIONS?',\n    'POLICIES AND PROTOCOLS FOR SCREENING AND TESTING?',\n    'POLICIES TO MITIGATE THE EFFECTS ON SUPPLIES ASSOCIATED WITH MASS TESTING, INCLUDING SWABS AND REAGENTS?',\n    'TECHNOLOGY ROADMAP FOR DIAGNOSTICS?',\n    'BARRIERS TO DEVELOPING AND SCALING UP NEW DIAGNOSTIC TESTS (E.G., MARKET FORCES), HOW FUTURE COALITION AND ACCELERATOR MODELS (E.G., COALITION FOR EPIDEMIC PREPAREDNESS INNOVATIONS) COULD PROVIDE CRITICAL FUNDING FOR DIAGNOSTICS, AND OPPORTUNITIES FOR A STREAMLINED REGULATORY ENVIRONMENT?',\n    'NEW PLATFORMS AND TECHNOLOGY (E.G., CRISPR) TO IMPROVE RESPONSE TIMES AND EMPLOY MORE HOLISTIC APPROACHES TO COVID-19 AND FUTURE DISEASES?',\n    'COUPLING GENOMICS AND DIAGNOSTIC TESTING ON A LARGE SCALE?',\n    'ENHANCE CAPABILITIES FOR RAPID SEQUENCING AND BIOINFORMATICS TO TARGET REGIONS OF THE GENOME THAT WILL ALLOW SPECIFICITY FOR A PARTICULAR VARIANT?',\n    'ENHANCE CAPACITY (PEOPLE, TECHNOLOGY, DATA) FOR SEQUENCING WITH ADVANCED ANALYTICS FOR UNKNOWN PATHOGENS, AND EXPLORE CAPABILITIES FOR DISTINGUISHING NATURALLY-OCCURRING PATHOGENS FROM INTENTIONAL?',\n    'ONE HEALTH SURVEILLANCE OF HUMANS AND POTENTIAL SOURCES OF FUTURE SPILLOVER OR ONGOING EXPOSURE FOR THIS ORGANISM AND FUTURE PATHOGENS, INCLUDING BOTH EVOLUTIONARY HOSTS (E.G., BATS) AND TRANSMISSION HOSTS (E.G., HEAVILY TRAFFICKED AND FARMED WILDLIFE AND DOMESTIC FOOD AND COMPANION SPECIES), INCLUSIVE OF ENVIRONMENTAL, DEMOGRAPHIC, AND OCCUPATIONAL RISK FACTORS?'\n]\n###\nquestions_specific = [questions_0_specific, questions_1_specific, questions_2_specific, questions_3_specific, questions_4_specific, questions_5_specific, questions_6_specific, questions_7_specific, questions_8_specific, questions_9_specific]","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**function 1: fct_create_dict_of_papers() ↓**\n\nwalk given path for papers.\nif conditions are met, then write paper paths to a main papers file.\n\nthis creates -\npapers.biorxiv_medrxiv.json,\npapers.comm_use_subset.json,\npapers.noncomm_use_subset.json,\npapers.pmc_custom_license.json.\n\nsee https://github.com/gisblog/nih-covid19/tree/master/covid19/kaggle/working.\n\ne.g. papers.biorxiv_medrxiv.json -\n> {\n>   \"paper\": [\n>     \"/kaggle/input/CORD-19-research-challenge/biorxiv_medrxiv/biorxiv_medrxiv/pdf_json/4602afcb8d95ebd9da583124384fd74299d20f5b.json\",...\n>     ]\n> }"},{"metadata":{"trusted":true},"cell_type":"code","source":"def fct_create_dict_of_papers(input_dir, input_path_of_papers, input_type_of_papers='json'): # see glob\n    dict_of_papers = {}\n    dict_of_papers['paper'] = []\n    dict_file = '/kaggle/working/papers.' + input_dir + '.json'\n    os.makedirs(os.path.dirname(dict_file), exist_ok=True)\n    with open(dict_file, 'w') as open_file:\n        for dirname, _, filenames in os.walk(input_path_of_papers):\n            for filename in filenames:\n                dir_file = os.path.join(dirname, filename)\n                if (filename).endswith(input_type_of_papers) and re.search('^(?=.*[0-9])(?=.*[a-z])([a-z0-9]+)$', filename.split('.')[0]) and ('/' + input_dir) in dirname: # comm/noncomm # regex: 860ffcab2771f1935ac5a59e986d416a603b3de3\n                    dict_of_papers['paper'].extend([dir_file])\n                break # test for kaggle\n        json.dump(dict_of_papers, open_file, indent=2, separators=(',', ': '))\n    print('*** fct_create_dict_of_papers ' + str(datetime.now()) + ' ***')\n    # return dict_of_papers # test\n\n### multiprocessing by dir - /biorxiv_medrxiv, /comm_use_subset, /noncomm_use_subset, /pmc_custom_license: v. multithreading ###\n# print('parent process id: ', os.getppid())\n# print('child process id: ', os.getpid())\n\nprocess_0_create_dict_of_papers = multiprocessing.Process(target=fct_create_dict_of_papers, args=('biorxiv_medrxiv', '/kaggle/input/CORD-19-research-challenge', 'json')) # args=(,) so it is a tuple if there is only 1 args\n# process_1_create_dict_of_papers = multiprocessing.Process(target=fct_create_dict_of_papers, args=('comm_use_subset', '/kaggle/input/CORD-19-research-challenge', 'json'))\n# process_2_create_dict_of_papers = multiprocessing.Process(target=fct_create_dict_of_papers, args=('noncomm_use_subset', '/kaggle/input/CORD-19-research-challenge', 'json'))\n# process_3_create_dict_of_papers = multiprocessing.Process(target=fct_create_dict_of_papers, args=('pmc_custom_license', '/kaggle/input/CORD-19-research-challenge', 'json'))\n\n### execute processes in parallel ###\nprocess_0_create_dict_of_papers.start()\n# process_1_create_dict_of_papers.start()\n# process_2_create_dict_of_papers.start()\n# process_3_create_dict_of_papers.start()\n\n### join processes back to the parent process (this) ###\nprocess_0_create_dict_of_papers.join()\n# process_1_create_dict_of_papers.join()\n# process_2_create_dict_of_papers.join()\n# process_3_create_dict_of_papers.join()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**function 2: fct_get_matches_from_papers ↓**\n\ngiven a list of papers in the main file, write the top unique answers to a task's QUESTIONS by evaluating the papers. for each paper, do -\n* collect each QUESTION and relevant elements from the paper.\n* convert the collection of words to vectors.\n* calculate euclidean distances from the QUESTIONS to the paper elements, and sort by similarity/distance.\n\nthis returns matches. \n\nscikit-learn -\n    \nword2vec is a group of models that represents each word in a large text as a vector in a space of n-dimensions (or features) making similar words closer to each other.\n\nthe euclidean distance (or cosine similarity) between 2 word vectors provided an effective method for measuring the linguistic/semantic similarity of the 2 words. nearest neighbor reveals relevant similarities outside an average vocabulary. similarity metrics used in nearest neighbor evaluations produce 1 scalar that quantifies the relatedness of its 2 words. this simplicity can be an issue since 2 words may exhibit other relationships. in order to capture that in a quantitative way, it is necessary to associate more than 1 number to a word pair.\n\npopulating glove required 1 pass through the entire covid19 dataset. for the large covid19 dataset, this pass was computationally expensive. subsequent training iterations would have been faster. also, pre-trained word vector datasets downloaded (eg wikipedia 2014 + gigaword 5) didn't match the semantics for covid19. # todo: train word vectors on covid19 corpus.\n\n    input layer | hidden layer | output layer | target layer\n    word | weight updated by training | neighbor word\n\n> class sklearn.feature_extraction.text.CountVectorizer(input='content', encoding='utf-8', decode_error='strict', strip_accents=None, lowercase=True, preprocessor=None, tokenizer=None, stop_words=None, token_pattern='(?u)\\b\\w\\w+\\b', ngram_range=(1, 1), analyzer='word', max_df=1.0, min_df=1, max_features=None, vocabulary=None, binary=False, dtype=<class 'numpy.int64'>)\n\ndefinitions below for any further tuning required  -\n\n* vocabulary: mapping, e.g. {dict}, where keys/terms:values/indices are in the feature matrix.\n- if not given, then determined from input docs.\n* document-frequency (df): ignored, if vocabulary != None.\n- ignore terms that have a min_df/max_df lower/higher than given:\n- if float, represents a proportion of docs. if integer, represent absolute counts.\n-- max_df = 0.50 means \"ignore terms that appear in more than 50% of the docs\"\n-- max_df = 25 means \"ignore terms that appear in more than 25 docs\"\n-- min_df = 0.01 means \"ignore terms that appear in less than 1% of the docs\"\n-- min_df = 5 means \"ignore terms that appear in less than 5 docs\"\n* analyzer: word or char/char_wb n-grams.\n- n-gram = contiguous sequence of n items from a given sample of text.\n- eg for sample \"to be or not to be\", 1-gram: [to, be, or, not, to, be], 2-gram: [to be, be or, or not, not to, to be].\n* ngram_range: unigrams (n=1, bag of words) can't capture phrases/multi-word expressions, and disregards word order dependence.\n- bag of words (bow) model doesn’t account for misspellings/word derivations/stemmings either.\n- so instead of unigrams, use bigrams (n=2) where occurrences of pairs of consecutive words are counted.\n- if analyzer isn't callable.\n* stop_words: if 'None', max_df used. if 'english', built-in stop word list used. if analyzer='word', [list] used.\n- stop_words_ attribute can get large and increase the model size when pickling ie converting obj (list, dict) to character stream."},{"metadata":{"trusted":true,"scrolled":true},"cell_type":"code","source":"def fct_get_matches_from_papers(input_path_to_paper, input_question, input_min_df=0.1, input_max_df=0.9): # see DataFrame\n    pp = pprint.PrettyPrinter()\n    paper = []\n    '''\n    each paper/item has the following structure:\n    {\n        \"paper_id\": \"...\", *\n        \"metadata\": {...}, *\n        \"abstract\": [...], *\n        \"body_text\": [...], *\n        \"bib_entries\": {...},\n        \"ref_entries\": {...},\n        \"back_matter\": [...]\n    }\n    '''\n    with open(input_path_to_paper) as paper_item: # todo: KeyError\n        # v. json.loads() doesn't take the file path, but the file contents as a string\n        paper_item_json_load = json.load(paper_item)\n        ### question: append [1, 2, 3, [4, 5]] v. extend [1, 2, 3, 4, 5] ###\n        paper.extend([input_question])\n        ### id ###\n        paper_id = paper_item_json_load['paper_id']\n        ### title, authors ###\n        metadata = paper_item_json_load['metadata']\n        ### abstract, citation/reference spans ###\n        abstract = paper_item_json_load['abstract']\n        for index, item in enumerate(abstract):\n            paper.extend([abstract[index]['text']])\n            ### text, citation/reference spans ###\n            body_text = paper_item_json_load['body_text']\n            for index, item in enumerate(body_text):\n                paper.extend([body_text[index]['text']])\n                ### bibliography ###\n                # bib_entries = paper_item_json_load['bib_entries']\n                ### findings ###\n                # ref_entries = paper_item_json_load['ref_entries']\n                ### funding, conflict of interest ###\n                # back_matter = paper_item_json_load['back_matter']\n    # pp.pprint(paper) # test: ['WHAT HAVE WE LEARNED ABOUT INFECTION PREVENTION AND CONTROL?',...]\n    \n    paper_match = []\n    try:\n        ### min_df, max_df and ngram_range ###\n        vectorizer = CountVectorizer(vocabulary=None, min_df=input_min_df, max_df=input_max_df, analyzer='word', ngram_range=(2, 3), stop_words=None)\n        ### learn vocabulary dict and return term-document matrix ###    \n        features = vectorizer.fit_transform(paper).todense() # todense() returns matrix v. toarray() returns ndarray\n        # print(vectorizer.vocabulary_) # test: {'dock8 deficient': 2,...}\n        ### for feature in features ###\n        for index, item in enumerate(features):\n            # print(euclidean_distances(features[0], features[i]), paper[i]) # test: [[0.]] WHAT HAVE WE LEARNED ABOUT INFECTION PREVENTION AND CONTROL?\n            paper_match.append([euclidean_distances(features[0], features[index])[0][0], paper[index]])\n        # print(paper_match) # test: [[0.0, 'WHAT HAVE WE LEARNED ABOUT INFECTION PREVENTION AND CONTROL?'],...]\n    except ValueError: # eg papers may req diff min_df or max_df\n        ### min_df, max_df and ngram_range ###\n        vectorizer = CountVectorizer(vocabulary=None, min_df=0.0, max_df=1.0, analyzer='word', ngram_range=(2, 3), stop_words=None)\n        ### learn vocabulary dict and return term-document matrix ###    \n        features = vectorizer.fit_transform(paper).todense()\n        ### for feature in features ###\n        for index, item in enumerate(features):\n            paper_match.append([euclidean_distances(features[0], features[index])[0][0], paper[index]])\n    print('*** fct_get_matches_from_papers ' + str(datetime.now()) + ' ***')\n    return sorted(paper_match, key=lambda k: k[0]) # returned vars don't get garbaged + remain accessible after the fct()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**function 3: fct_get_answer_from_matches() ↓**\n\nfrom the sorting, remove the original QUESTIONS and return the top unique answers."},{"metadata":{"trusted":true},"cell_type":"code","source":"def fct_get_answer_from_matches(input_path_to_paper, input_min_df, input_max_df, input_task=0, input_top=4):\n    pp = pprint.PrettyPrinter()\n    answer = []\n    for index in range(input_top):\n        try:\n            if (fct_get_matches_from_papers(input_path_to_paper, questions[input_task], input_min_df, input_max_df)[index][0] != 0.0): # remove original question (+ no perfect answer with score = 0.0?)\n                answer.append(fct_get_matches_from_papers(input_path_to_paper, questions[input_task], input_min_df, input_max_df)[index][1]) # (question)[row(distance, answer)][answer]\n        except IndexError:\n            pass\n    for question in questions_detail[input_task]:\n        ### get paper/text: if euclidean distance < #, may return empty, hence top 3 (after removing original question) ###\n        # print(question) # test: WHAT IS KNOWN ABOUT TRANSMISSION, INCUBATION, AND ENVIRONMENTAL STABILITY?\n        for index in range(input_top):\n            try:\n                if (fct_get_matches_from_papers(input_path_to_paper, question, input_min_df, input_max_df)[index][0] != 0.0):\n                    answer.append(fct_get_matches_from_papers(input_path_to_paper, question, input_min_df, input_max_df)[index][1])\n            except IndexError:\n                pass\n    for question in questions_specific[input_task]:\n        ### get paper/text ###\n        for index in range(input_top):\n            try:\n                if (fct_get_matches_from_papers(input_path_to_paper, question, input_min_df, input_max_df)[index][0] != 0.0):\n                    answer.append(fct_get_matches_from_papers(input_path_to_paper, question, input_min_df, input_max_df)[index][1])\n            except IndexError:\n                pass\n    answer_unique = list(dict.fromkeys(answer)) # also remove duplicate distances\n    # pp.pprint(answer_unique) # test\n    print('*** fct_get_answer_from_matches ' + str(datetime.now()) + ' ***')\n    return answer_unique","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**function 4: fct_write_answers() ↓**\n\nwrite the answers to an answer file.\n\nthis creates all the answer files, following the same storage structure as the input data.\n\nsee https://github.com/gisblog/nih-covid19/tree/master/covid19/kaggle/working/CORD-19-research-challenge/2020-03-13.\n\ne.g. 3c70c99afc7a38df3c4807857856ea258d378429.json -\n\n> {\n  > \"paper_id\": \"3c70c99afc7a38df3c4807857856ea258d378429\",\n  > \"task\": 0,\n  > \"abstract\": \"WHAT IS KNOWN ABOUT TRANSMISSION, INCUBATION, AND ENVIRONMENTAL STABILITY?\",\n  > \"body_text\": [\n    > \"Introduction encompasses residues 347-726 and has the characteristic (\\u03b2/\\u03b1) 8 topology found in the catalytic 109 domain of family 31 glycoside hydrolases. C-terminal to the catalytic (\\u03b2/\\u03b1) 8 domain, the GAA 110 structure contains two additional \\u03b2-sandwich domains. Domain C1 spans residues 727-820 and 111 is composed of seven antiparallel \\u03b2-strands in 5-and 2-stranded \\u03b2-sheets. The C2 domain has 112 a \\u03b2-sandwich topology and is composed of 10 \\u03b2-strands in two 5-stranded \\u03b2-sheets, with two 113 small \\u03b1-helices inserted ( Fig. 1 and Fig. 1 supplement B) . and the surface-exposed C938 and C952 in the C2 domain (Fig. 1) .\",\n    > \"Pompe disease results from a defect in human acid \\u03b1-glucosidase (GAA), a lysosomal enzyme 17 that cleaves terminal \\u03b11-4 and \\u03b11-6 glucose from glycogen. In Pompe disease (also known as 18 Glycogen Storage Disorder type II), the accumulation of undegraded glycogen in lysosomes 19 leads to cellular dysfunction, primarily in muscle and heart tissues. Pompe disease is an active 20 candidate of clinical research, with pharmacological chaperone therapy tested and enzyme 21 replacement therapy approved. Despite production of large amounts of recombinant GAA 22 annually, the structure of GAA has not been reported until now. Here, we describe the first 23 structure of GAA, at 1.7\\u00c5 resolution. Three structures of GAA complexes reveal the molecular 24 basis for the hundreds of mutations that lead to Pompe disease and for pharmacological 25 chaperoning in the protein. The GAA structure reveals a surprising second sugar-binding site 26\",\n    > ...\n  > ]\n> }"},{"metadata":{"trusted":true,"scrolled":true},"cell_type":"code","source":"def fct_write_answers(input_path_of_papers, input_min_df, input_max_df, input_task, input_top):\n    with open(input_path_of_papers) as open_file:\n        json_file = json.load(open_file)\n        for index, item in enumerate(json_file['paper']):\n            print('*** ' + item + ' ' + str(datetime.now()) + ' ***')\n            ### id ###\n            paper_file = item.split('/')[-1]\n            paper_id = paper_file.split('.')[0]\n            # todo: add title\n            ### json ###\n            answer_json = {\n                \"paper_id\": paper_id,\n                \"task\": input_task,\n                \"abstract\": questions[input_task],\n                \"body_text\": fct_get_answer_from_matches(item, input_min_df, input_max_df, input_task, input_top) # item = input_path_to_paper\n            }\n            output_answer_file = '/kaggle/working' + item\n            os.makedirs(os.path.dirname(output_answer_file), exist_ok=True)\n            with open(output_answer_file, 'w') as open_answer:\n                json.dump(answer_json, open_answer, indent=2, separators=(',', ': '))\n    print('*** fct_write_answers ' + str(datetime.now()) + ' ***')\n\n### multiprocessing ###\nprocess_0_write_answers = multiprocessing.Process(target=fct_write_answers, args=('/kaggle/working/papers.biorxiv_medrxiv.json', 0.1, 0.9, 0, 4))\n# process_1_write_answers = multiprocessing.Process(target=fct_write_answers, args=('/kaggle/working/papers.comm_use_subset.json', 0.1, 0.9, 0, 4))\n# process_2_write_answers = multiprocessing.Process(target=fct_write_answers, args=('/kaggle/working/papers.noncomm_use_subset.json', 0.1, 0.9, 0, 4))\n# process_3_write_answers = multiprocessing.Process(target=fct_write_answers, args=('/kaggle/working/papers.pmc_custom_license.json', 0.1, 0.9, 0, 4))\n\n### execute processes in parallel ###\nprocess_0_write_answers.start()\n# process_1_write_answers.start()\n# process_2_write_answers.start()\n# process_3_write_answers.start()\n\n### join processes back to the parent process (this) ###\nprocess_0_write_answers.join()\n# process_1_write_answers.join()\n# process_2_write_answers.join()\n# process_3_write_answers.join()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**function 5: fct_merge_answers() ↓**\n\nwalk given path for answer files.\nif conditions are met, then merge all answers on a given path by task # and source type into a main answer file.\nthe merged JSON is structured like the original papers, and contains pointers to the original papers for reference.\n\nthis creates -\nanswers.task.0.biorxiv_medrxiv.json,\nanswers.task.0.comm_use_subset.json,\nanswers.task.0.noncomm_use_subset.json,\nanswers.task.0.pmc_custom_license.json.\n\nsee https://github.com/gisblog/nih-covid19/tree/master/covid19/kaggle/working.\n\ne.g. answers.task.0.comm_use_subset.json -\n> [\n> ...\n> {\n  > \"paper_id\": \"fffaed7e9353b7df6c4ca8f66b62e117013cb86d\",\n  > \"task\": 0,\n  > \"abstract\": \"WHAT IS KNOWN ABOUT TRANSMISSION, INCUBATION, AND ENVIRONMENTAL STABILITY?\",\n  > \"body_text\": [\n    > \"N-glycosylation on both E and NS1 proteins has been shown to play important roles throughout the DENV infection cycle from virion attachment, entry, maturation, assembly to secretion.\",\n    > \"SY, TN-K, and SA wrote the manuscript. PR provided suggestions and edited the manuscript.\"\n  > ]\n> }\n> ...\n> ]"},{"metadata":{"trusted":true,"scrolled":true},"cell_type":"code","source":"def fct_merge_answers(input_task, input_dir, input_path_of_answers, input_type_of_answers='json'):\n    answer_file = '/kaggle/working/answers.task.' + str(input_task) + '.' + input_dir + '.json'\n    os.makedirs(os.path.dirname(answer_file), exist_ok=True)\n    with open(answer_file, 'w') as open_file:\n        open_file.write('[')\n        for dirname, _, filenames in os.walk(input_path_of_answers):\n            for filename in filenames:\n                dir_file = os.path.join(dirname, filename)\n                if (filename).endswith(input_type_of_answers) and re.search('^(?=.*[0-9])(?=.*[a-z])([a-z0-9]+)$', filename.split('.')[0]) and ('/' + input_dir) in dirname:\n                    with open(dir_file) as answer_item:\n                        print('*** ' + dir_file + ' ' + str(datetime.now()) + ' ***')\n                        '''\n                        {\n                          \"paper_id\": \"...\",\n                          \"task\": 0,\n                          \"abstract\": \"...?\",\n                          \"body_text\": [\"...\"]\n                        }\n                        '''\n                        answer_item_json_load = json.load(answer_item)\n                        try:\n                            if (answer_item_json_load['task'] == input_task):\n                                json.dump(answer_item_json_load, open_file, indent=2, separators=(',', ': '))\n                        except KeyError:\n                            json.dump({}, open_file, indent=2, separators=(',', ': '))\n                        if (filename != filenames[-1]):\n                            open_file.write(',')\n        open_file.write(']')\n    print('*** fct_merge_answers ' + str(datetime.now()) + ' ***')\n\n### multiprocessing ###\nprocess_0_merge_answers = multiprocessing.Process(target=fct_merge_answers, args=(0, 'biorxiv_medrxiv', '/kaggle/working/CORD-19-research-challenge/2020-03-13/biorxiv_medrxiv/biorxiv_medrxiv/', 'json'))\n# process_1_merge_answers = multiprocessing.Process(target=fct_merge_answers, args=(0, 'comm_use_subset', '/kaggle/working/CORD-19-research-challenge/2020-03-13/comm_use_subset/comm_use_subset/', 'json'))\n# process_2_merge_answers = multiprocessing.Process(target=fct_merge_answers, args=(0, 'noncomm_use_subset', '/kaggle/working/CORD-19-research-challenge/2020-03-13/noncomm_use_subset/noncomm_use_subset/', 'json'))\n# process_3_merge_answers = multiprocessing.Process(target=fct_merge_answers, args=(0, 'pmc_custom_license', '/kaggle/working/CORD-19-research-challenge/2020-03-13/pmc_custom_license/pmc_custom_license/', 'json'))\n\n### execute processes in parallel ###\nprocess_0_merge_answers.start()\n# process_1_merge_answers.start()\n# process_2_merge_answers.start()\n# process_3_merge_answers.start()\n\n### join processes back to the parent process (this) ###\nprocess_0_merge_answers.join()\n# process_1_merge_answers.join()\n# process_2_merge_answers.join()\n# process_3_merge_answers.join()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"see files at https://github.com/gisblog/nih-covid19.\n\nfor e.g., potential answers for task 1 - \"WHAT IS KNOWN ABOUT TRANSMISSION, INCUBATION, AND ENVIRONMENTAL STABILITY?\" can be found here: (broken down by source type)\n* <a href=\"//raw.githubusercontent.com/gisblog/nih-covid19/master/covid19/kaggle/working/answers.task.0.biorxiv_medrxiv.json\">bioRxiv-medRxiv</a> (600 kb)\n* <a href=\"//raw.githubusercontent.com/gisblog/nih-covid19/master/covid19/kaggle/working/answers.task.0.comm_use_subset.json\">Commmercial Use</a> (9.2 mb)\n* <a href=\"//raw.githubusercontent.com/gisblog/nih-covid19/master/covid19/kaggle/working/answers.task.0.noncomm_use_subset.json\">Non-commercial Use</a> (1.7 mb)\n* <a href=\"//raw.githubusercontent.com/gisblog/nih-covid19/master/covid19/kaggle/working/answers.task.0.pmc_custom_license.json\">PubMed Central (PMC)</a> (1.2 mb)\n\npotential answers for task 3 - \"HELP US UNDERSTAND HOW GEOGRAPHY AFFECTS VIRALITY?\" can be found here:\n* <a href=\"//raw.githubusercontent.com/gisblog/nih-covid19/master/covid19/kaggle/working/answers.task.3.biorxiv_medrxiv.json\">bioRxiv-medRxiv</a> (660 kb)\n* <a href=\"//raw.githubusercontent.com/gisblog/nih-covid19/master/covid19/kaggle/working/answers.task.3.comm_use_subset.json\">Commmercial Use</a> (9.1 mb)\n* <a href=\"//raw.githubusercontent.com/gisblog/nih-covid19/master/covid19/kaggle/working/answers.task.3.noncomm_use_subset.json\">Non-commercial Use</a> (1.7 mb)\n* <a href=\"//raw.githubusercontent.com/gisblog/nih-covid19/master/covid19/kaggle/working/answers.task.3.pmc_custom_license.json\">PubMed Central (PMC)</a> (1.2 mb)\n\nwe can also geoparse and geocode the extracted NLP answers for visualization - see <a href=\"//github.com/gisblog/nih-covid19/blob/master/covid19/kaggle/working/kaggle-geo.py\">kaggle-geo.py</a>.\n\ngeoparsed answers for task 3 - \"HELP US UNDERSTAND HOW GEOGRAPHY AFFECTS VIRALITY?\":\n\ncities -\n\n* <a href=\"//raw.githubusercontent.com/gisblog/nih-covid19/master/covid19/kaggle/working/cities.answers.task.3.biorxiv_medrxiv.json\">bioRxiv-medRxiv</a>\n* <a href=\"//raw.githubusercontent.com/gisblog/nih-covid19/master/covid19/kaggle/working/cities.answers.task.3.comm_use_subset.json\">Commmercial Use</a>\n* <a href=\"//raw.githubusercontent.com/gisblog/nih-covid19/master/covid19/kaggle/working/cities.answers.task.3.noncomm_use_subset.json\">Non-commercial Use</a>\n* <a href=\"//raw.githubusercontent.com/gisblog/nih-covid19/master/covid19/kaggle/working/cities.answers.task.3.pmc_custom_license.json\">PubMed Central (PMC)</a>\n\ncountry codes and counts -\n\n* <a href=\"//raw.githubusercontent.com/gisblog/nih-covid19/master/covid19/kaggle/working/country_mentions.answers.task.3.biorxiv_medrxiv.json\">bioRxiv-medRxiv</a>\n* <a href=\"//raw.githubusercontent.com/gisblog/nih-covid19/master/covid19/kaggle/working/country_mentions.answers.task.3.comm_use_subset.json\">Commmercial Use</a>\n* <a href=\"//raw.githubusercontent.com/gisblog/nih-covid19/master/covid19/kaggle/working/country_mentions.answers.task.3.noncomm_use_subset.json\">Non-commercial Use</a>\n* <a href=\"//raw.githubusercontent.com/gisblog/nih-covid19/master/covid19/kaggle/working/country_mentions.answers.task.3.pmc_custom_license.json\">PubMed Central (PMC)</a>"}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}