{"cells":[{"metadata":{},"cell_type":"markdown","source":"## About this notebook\nThe helper functions are taken from here : https://www.kaggle.com/xhlulu/cord-19-eda-parse-json-and-generate-clean-csv\n(Many thanks to @xhlulu for the convenient pre-treatment)\n\nThe goal of this notebook is to present a methodology to find match between the articles and the tasks descriptions, and to have a visual representation of the rlationship between tasks and articles\n\n### Summary\n\n#### 1. [Loading data](#helper)\nUsing @xhlulu's work \n#### 2. [Doc2Vec training on abstracts data](#train)\nTrain Gensim Doc2Vec model on abstracts \n#### 2. [Doc2Vec projection of tasks](#tasks)  \n   Use the model to convert tasks and abstracts to vectors\n#### 3. [Nearest Neighbors analysis](#search)\n   Find which abstracts are closest to each task description\n#### 4. [Clustering and PCA + UMAP to visualise abstract and texts in a 3D space](#clustering)\n   Use the vectors to perform a clustering (K-Means) and then use PCA + UMAP to reduce dimension to 3 and visualize the results \n\n\n\n\nFind out more on this blog post : https://www.quantmetry.com/kaggle-covid-19-open-research-dataset-challenge-cord-19/"},{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true},"cell_type":"code","source":"import os\nimport json\nfrom pprint import pprint\nfrom copy import deepcopy\n\nimport numpy as np\nimport pandas as pd\nfrom tqdm.notebook import tqdm","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"<a id='helper'></a>\n## 1. Load Data (section by @xhulu)"},{"metadata":{},"cell_type":"markdown","source":"Unhide the cell below to find the definition of the following functions:\n* `format_name(author)`\n* `format_affiliation(affiliation)`\n* `format_authors(authors, with_affiliation=False)`\n* `format_body(body_text)`\n* `format_bib(bibs)`"},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":"def format_name(author):\n    middle_name = \" \".join(author['middle'])\n    \n    if author['middle']:\n        return \" \".join([author['first'], middle_name, author['last']])\n    else:\n        return \" \".join([author['first'], author['last']])\n\n\ndef format_affiliation(affiliation):\n    text = []\n    location = affiliation.get('location')\n    if location:\n        text.extend(list(affiliation['location'].values()))\n    \n    institution = affiliation.get('institution')\n    if institution:\n        text = [institution] + text\n    return \", \".join(text)\n\ndef format_authors(authors, with_affiliation=False):\n    name_ls = []\n    \n    for author in authors:\n        name = format_name(author)\n        if with_affiliation:\n            affiliation = format_affiliation(author['affiliation'])\n            if affiliation:\n                name_ls.append(f\"{name} ({affiliation})\")\n            else:\n                name_ls.append(name)\n        else:\n            name_ls.append(name)\n    \n    return \", \".join(name_ls)\n\ndef format_body(body_text):\n    texts = [(di['section'], di['text']) for di in body_text]\n    texts_di = {di['section']: \"\" for di in body_text}\n    \n    for section, text in texts:\n        texts_di[section] += text\n\n    body = \"\"\n\n    for section, text in texts_di.items():\n        #body += section\n        #body += \"\\n\\n\"\n        body += text\n        #body += \"\\n\\n\"\n    \n    return body\n\ndef format_bib(bibs):\n    if type(bibs) == dict:\n        bibs = list(bibs.values())\n    bibs = deepcopy(bibs)\n    formatted = []\n    \n    for bib in bibs:\n        bib['authors'] = format_authors(\n            bib['authors'], \n            with_affiliation=False\n        )\n        formatted_ls = [str(bib[k]) for k in ['title', 'authors', 'venue', 'year']]\n        formatted.append(\", \".join(formatted_ls))\n\n    return \"; \".join(formatted)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"Unhide the cell below to find the definition of the following functions:\n* `load_files(dirname)`\n* `generate_clean_df(all_files)`"},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":"def load_files(dirname):\n    filenames = os.listdir(dirname)\n    raw_files = []\n\n    for filename in tqdm(filenames):\n        filename = dirname + filename\n        file = json.load(open(filename, 'rb'))\n        raw_files.append(file)\n    \n    return raw_files\n\ndef generate_clean_df(all_files):\n    cleaned_files = []\n    \n    for file in tqdm(all_files):\n        features = [\n            file['paper_id'],\n            file['metadata']['title'],\n            format_authors(file['metadata']['authors']),\n            format_authors(file['metadata']['authors'], \n                           with_affiliation=True),\n            format_body(file['abstract']),\n            format_body(file['body_text']),\n            format_bib(file['bib_entries']),\n            file['metadata']['authors'],\n            file['bib_entries']\n        ]\n\n        cleaned_files.append(features)\n\n    col_names = ['paper_id', 'title', 'authors',\n                 'affiliations', 'abstract', 'text', \n                 'bibliography','raw_authors','raw_bibliography']\n\n    clean_df = pd.DataFrame(cleaned_files, columns=col_names)\n    clean_df.head()\n    \n    return clean_df","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### Biorxiv: Generate CSV\n\nIn this section, I show you how to manually generate the CSV files. As you can see, it's now super simple because of the `format_` helper functions. In the next sections, I show you have to generate them in 3 lines using the `load_files` and `generate_clean_dr` helper functions."},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":"biorxiv_dir = '/kaggle/input/CORD-19-research-challenge/biorxiv_medrxiv/biorxiv_medrxiv/pdf_json/'\nfilenames = os.listdir(biorxiv_dir)\nprint(\"Number of articles retrieved from biorxiv:\", len(filenames))\n\nall_files = []\n\nfor filename in filenames:\n    filename = biorxiv_dir + filename\n    file = json.load(open(filename, 'rb'))\n    all_files.append(file)\n    \ncleaned_files = []\n\nfor file in tqdm(all_files):\n    features = [\n        file['paper_id'],\n        file['metadata']['title'],\n        format_authors(file['metadata']['authors']),\n        format_authors(file['metadata']['authors'], \n                       with_affiliation=True),\n        format_body(file['abstract']),\n        format_body(file['body_text']),\n        format_bib(file['bib_entries']),\n        file['metadata']['authors'],\n        file['bib_entries']\n    ]\n    \n    cleaned_files.append(features)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":"col_names = [\n    'paper_id', \n    'title', \n    'authors',\n    'affiliations', \n    'abstract', \n    'text', \n    'bibliography',\n    'raw_authors',\n    'raw_bibliography'\n]\n\nclean_df = pd.DataFrame(cleaned_files, columns=col_names)\nclean_df.head(2)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":"pmc_dir = '/kaggle/input/CORD-19-research-challenge/custom_license/custom_license/pdf_json/'\npmc_files = load_files(pmc_dir)\npmc_df = generate_clean_df(pmc_files)\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":"comm_dir = '/kaggle/input/CORD-19-research-challenge/comm_use_subset/comm_use_subset/pdf_json/'\ncomm_files = load_files(comm_dir)\ncomm_df = generate_clean_df(comm_files)\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":"noncomm_dir = '/kaggle/input/CORD-19-research-challenge/noncomm_use_subset/noncomm_use_subset/pdf_json/'\nnoncomm_files = load_files(noncomm_dir)\nnoncomm_df = generate_clean_df(noncomm_files)\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"complete_df = pd.concat([clean_df, pmc_df, comm_df, noncomm_df])\ncomplete_df.shape","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"Keep only texts with minimal number of words using regex"},{"metadata":{"trusted":true},"cell_type":"code","source":"import re\ncomplete_df = complete_df[complete_df['text'].apply(lambda x: len(re.findall(r\"(?i)\\b[a-z]+\\b\", x))) > 1000]                                           \ncomplete_df.shape","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"complete_df.reset_index(inplace=True, drop=True)\ncomplete_df.iloc[42]['text'][:500]","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"<a id='train'></a>\n## 2. Train gensim Doc2Vec model\n\n\nWord2Vec is a very famous algorithm for word embeddings. You can find a very good explanation of this algorithm in [this amazing blog post](http://jalammar.github.io/illustrated-word2vec/)\nThe main idea is beautifully resumed in this quote by  [John Ruper Firth](https://en.wikipedia.org/wiki/John_Rupert_Firth) :\n\n *You shall know a word by the company it keeps*\n\nWord2Vec came up with this very innovative idea where you train a neural network on a task (finding a word given surrounding words - or the reverse) but you are not interested in the result of this task, rather in the weights of the network’s hidden layer, that will be used to derive the word vectors. \n\nDoc2Vec has an augmented objective to have one vector for a sequence of words. \nThe first idea to do that could be to just take the mean vector of the Word2Vec vectors of each word in the sequence, however the  Doc2Vec authors had a smarter idea ([see the paper here](https://arxiv.org/pdf/1405.4053v2.pdf))\n).\nIt follows the same principle as Word2Vec : training on a task to get the hidden layer as a vector, but this time, we add the whole paragraph matrix along the window words as an input to the classifier. See on the following figure the task for the distributed memory model (you can check the other algorithm : Distributed bag of words in the article) \n\nFortunately the famous NLP library Gensim comes with an optimized class to train the Doc2Vec model \n[We can follow this tutorial](https://radimrehurek.com/gensim/auto_examples/tutorials/run_doc2vec_lee.html#sphx-glr-auto-examples-tutorials-run-doc2vec-lee-py) on training Doc2Vec with gensim"},{"metadata":{"trusted":true},"cell_type":"code","source":"import gensim\n\ndef read_corpus(df, column, tokens_only=False):\n    \"\"\"\n    Arguments\n    ---------\n        df: pd.DataFrame\n        column: str \n            text column name\n        tokens_only: bool\n            wether to add tags or not\n    \"\"\"\n    for i, line in enumerate(df[column]):\n\n        tokens = gensim.parsing.preprocess_string(line)\n        if tokens_only:\n            yield tokens\n        else:\n            # For training data, add tags\n            yield gensim.models.doc2vec.TaggedDocument(tokens, [i])","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"We take a sample to allow the training to run into a reasonable time, frac_of_articles (between 0 and 1) should be set to 1 to  have more data to train the model "},{"metadata":{"trusted":true},"cell_type":"code","source":"import random\nfrac_of_articles = 1\ntrain_df  = complete_df.sample(frac=frac_of_articles, random_state=42)\ntrain_corpus = (list(read_corpus(train_df, 'abstract'))) ","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# using distributed memory model\nmodel = gensim.models.doc2vec.Doc2Vec(dm=1, vector_size=50, min_count=10, dm_mean=1, epochs=20, seed=42, workers=6)\nmodel.build_vocab(train_corpus)\nmodel.train(train_corpus, total_examples=model.corpus_count, epochs=model.epochs)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"<a id='tasks'></a>\n\n## 3. Tasks descriptions projection\n\nWe take the texts form all tasks and calculate their vectors with our Doc2Vec model so that we can find nearest neighbors"},{"metadata":{"trusted":true,"_kg_hide-input":true},"cell_type":"code","source":"task_1 = \"\"\"What is known about transmission, incubation, and environmental stability of COVID-19? What do we know about natural history, transmission, and diagnostics for the virus? What have we learned about infection prevention and control?\nRange of incubation periods for the disease in humans (and how this varies across age and health status) and how long individuals are contagious, even after recovery.\nPrevalence of asymptomatic shedding and transmission (e.g., particularly children).\nSeasonality of transmission.\nPhysical science of the coronavirus (e.g., charge distribution, adhesion to hydrophilic/phobic surfaces, environmental survival to inform decontamination efforts for affected areas and provide information about viral shedding).\nPersistence and stability on a multitude of substrates and sources (e.g., nasal discharge, sputum, urine, fecal matter, blood).\nPersistence of virus on surfaces of different materials (e,g., copper, stainless steel, plastic).\nNatural history of the virus and shedding of it from an infected person\nImplementation of diagnostics and products to improve clinical processes\nDisease models, including animal models for infection, disease and transmission\nTools and studies to monitor phenotypic change and potential adaptation of the virus\nImmune response and immunity\nEffectiveness of movement control strategies to prevent secondary transmission in health care and community settings\nEffectiveness of personal protective equipment (PPE) and its usefulness to reduce risk of transmission in health care and community settings\nRole of the environment in transmission\"\"\"\n\ntask_2 = \"\"\"What do we know about COVID-19 risk factors? What have we learned from epidemiological studies?\nData on potential risks factors\nSmoking, pre-existing pulmonary disease\nCo-infections (determine whether co-existing respiratory/viral infections make the virus more transmissible or virulent) and other co-morbidities\nNeonates and pregnant women\nSocio-economic and behavioral factors to understand the economic impact of the virus and whether there were differences.\nTransmission dynamics of the virus, including the basic reproductive number, incubation period, serial interval, modes of transmission and environmental factors\nSeverity of disease, including risk of fatality among symptomatic hospitalized patients, and high-risk patient groups\nSusceptibility of populations\nPublic health mitigation measures that could be effective for control\"\"\"\n\ntask_3 = \"\"\"What do we know about virus genetics, origin, and evolution? What do we know about the virus origin and management measures at the human-animal interface?\nReal-time tracking of whole genomes and a mechanism for coordinating the rapid dissemination of that information to inform the development of diagnostics and therapeutics and to track variations of the virus over time.\nAccess to geographic and temporal diverse sample sets to understand geographic distribution and genomic differences, and determine whether there is more than one strain in circulation. Multi-lateral agreements such as the Nagoya Protocol could be leveraged.\nEvidence that livestock could be infected (e.g., field surveillance, genetic sequencing, receptor binding) and serve as a reservoir after the epidemic appears to be over.\nEvidence of whether farmers are infected, and whether farmers could have played a role in the origin.\nSurveillance of mixed wildlife- livestock farms for SARS-CoV-2 and other coronaviruses in Southeast Asia.\nExperimental infections to test host range for this pathogen.\nAnimal host(s) and any evidence of continued spill-over to humans\nSocioeconomic and behavioral risk factors for this spill-over\nSustainable risk reduction strategies\"\"\"\n\ntask_4 = \"\"\"What do we know about vaccines and therapeutics? What has been published concerning research and development and evaluation efforts of vaccines and therapeutics?\nEffectiveness of drugs being developed and tried to treat COVID-19 patients.\nClinical and bench trials to investigate less common viral inhibitors against COVID-19 such as naproxen, clarithromycin, and minocyclinethat that may exert effects on viral replication.\nMethods evaluating potential complication of Antibody-Dependent Enhancement (ADE) in vaccine recipients.\nExploration of use of best animal models and their predictive value for a human vaccine.\nCapabilities to discover a therapeutic (not vaccine) for the disease, and clinical effectiveness studies to discover therapeutics, to include antiviral agents.\nAlternative models to aid decision makers in determining how to prioritize and distribute scarce, newly proven therapeutics as production ramps up. This could include identifying approaches for expanding production capacity to ensure equitable and timely distribution to populations in need.\nEfforts targeted at a universal coronavirus vaccine.\nEfforts to develop animal models and standardize challenge studies\nEfforts to develop prophylaxis clinical studies and prioritize in healthcare workers\nApproaches to evaluate risk for enhanced disease after vaccination\nAssays to evaluate vaccine immune response and process development for vaccines, alongside suitable animal models [in conjunction with therapeutics]\"\"\"\n\n\ntask_5 = \"\"\"What do we know about the effectiveness of non-pharmaceutical interventions? What is known about equity and barriers to compliance for non-pharmaceutical interventions?\nGuidance on ways to scale up NPIs in a more coordinated way (e.g., establish funding, infrastructure and authorities to support real time, authoritative (qualified participants) collaboration with all states to gain consensus on consistent guidance and to mobilize resources to geographic areas where critical shortfalls are identified) to give us time to enhance our health care delivery system capacity to respond to an increase in cases.\nRapid design and execution of experiments to examine and compare NPIs currently being implemented. DHS Centers for Excellence could potentially be leveraged to conduct these experiments.\nRapid assessment of the likely efficacy of school closures, travel bans, bans on mass gatherings of various sizes, and other social distancing approaches.\nMethods to control the spread in communities, barriers to compliance and how these vary among different populations..\nModels of potential interventions to predict costs and benefits that take account of such factors as race, income, disability, age, geographic location, immigration status, housing status, employment status, and health insurance status.\nPolicy changes necessary to enable the compliance of individuals with limited resources and the underserved with NPIs.\nResearch on why people fail to comply with public health advice, even if they want to do so (e.g., social or financial costs may be too high).\nResearch on the economic impact of this or any pandemic. This would include identifying policy and programmatic alternatives that lessen/mitigate risks to critical government services, food distribution and supplies, access to critical household supplies, and access to health diagnoses, treatment, and needed care, regardless of ability to pay.\"\"\"\n\ntask_6 = \"\"\"What do we know about diagnostics and surveillance? What has been published concerning systematic, holistic approach to diagnostics (from the public health surveillance perspective to being able to predict clinical outcomes)?\nHow widespread current exposure is to be able to make immediate policy recommendations on mitigation measures. Denominators for testing and a mechanism for rapidly sharing that information, including demographics, to the extent possible. Sampling methods to determine asymptomatic disease (e.g., use of serosurveys (such as convalescent samples) and early detection of disease (e.g., use of screening of neutralizing antibodies such as ELISAs).\nEfforts to increase capacity on existing diagnostic platforms and tap into existing surveillance platforms.\nRecruitment, support, and coordination of local expertise and capacity (public, private—commercial, and non-profit, including academic), including legal, ethical, communications, and operational issues.\nNational guidance and guidelines about best practices to states (e.g., how states might leverage universities and private laboratories for testing purposes, communications to public health officials and the public).\nDevelopment of a point-of-care test (like a rapid influenza test) and rapid bed-side tests, recognizing the tradeoffs between speed, accessibility, and accuracy.\nRapid design and execution of targeted surveillance experiments calling for all potential testers using PCR in a defined area to start testing and report to a specific entity. These experiments could aid in collecting longitudinal samples, which are critical to understanding the impact of ad hoc local interventions (which also need to be recorded).\nSeparation of assay development issues from instruments, and the role of the private sector to help quickly migrate assays onto those devices.\nEfforts to track the evolution of the virus (i.e., genetic drift or mutations) and avoid locking into specific reagents and surveillance/detection schemes.\nLatency issues and when there is sufficient viral load to detect the pathogen, and understanding of what is needed in terms of biological and environmental sampling.\nUse of diagnostics such as host response markers (e.g., cytokines) to detect early disease or predict severe disease progression, which would be important to understanding best clinical practice and efficacy of therapeutic interventions.\nPolicies and protocols for screening and testing.\nPolicies to mitigate the effects on supplies associated with mass testing, including swabs and reagents.\nTechnology roadmap for diagnostics.\nBarriers to developing and scaling up new diagnostic tests (e.g., market forces), how future coalition and accelerator models (e.g., Coalition for Epidemic Preparedness Innovations) could provide critical funding for diagnostics, and opportunities for a streamlined regulatory environment.\nNew platforms and technology (e.g., CRISPR) to improve response times and employ more holistic approaches to COVID-19 and future diseases.\nCoupling genomics and diagnostic testing on a large scale.\nEnhance capabilities for rapid sequencing and bioinformatics to target regions of the genome that will allow specificity for a particular variant.\nEnhance capacity (people, technology, data) for sequencing with advanced analytics for unknown pathogens, and explore capabilities for distinguishing naturally-occurring pathogens from intentional.\nOne Health surveillance of humans and potential sources of future spillover or ongoing exposure for this organism and future pathogens, including both evolutionary hosts (e.g., bats) and transmission hosts (e.g., heavily trafficked and farmed wildlife and domestic food and companion species), inclusive of environmental, demographic, and occupational risk factors.\n        \"\"\"\n\ntask_7 = \"\"\"What has been published about medical care? What has been published concerning surge capacity and nursing homes? What has been published concerning efforts to inform allocation of scarce resources? What do we know about personal protective equipment? What has been published concerning alternative methods to advise on disease management? What has been published concerning processes of care? What do we know about the clinical characterization and management of the virus?\nResources to support skilled nursing facilities and long term care facilities.\nMobilization of surge medical staff to address shortages in overwhelmed communities\nAge-adjusted mortality data for Acute Respiratory Distress Syndrome (ARDS) with/without other organ failure – particularly for viral etiologies\nExtracorporeal membrane oxygenation (ECMO) outcomes data of COVID-19 patients\nOutcomes data for COVID-19 after mechanical ventilation adjusted for age.\nKnowledge of the frequency, manifestations, and course of extrapulmonary manifestations of COVID-19, including, but not limited to, possible cardiomyopathy and cardiac arrest.\nApplication of regulatory standards (e.g., EUA, CLIA) and ability to adapt care to crisis standards of care level.\nApproaches for encouraging and facilitating the production of elastomeric respirators, which can save thousands of N95 masks.\nBest telemedicine practices, barriers and faciitators, and specific actions to remove/expand them within and across state boundaries.\nGuidance on the simple things people can do at home to take care of sick people and manage disease.\nOral medications that might potentially work.\nUse of AI in real-time health care delivery to evaluate interventions, risk factors, and outcomes in a way that could not be done manually.\nBest practices and critical challenges and innovative solutions and technologies in hospital flow and organization, workforce protection, workforce allocation, community-based support resources, payment, and supply chain management to enhance capacity, efficiency, and outcomes.\nEfforts to define the natural history of disease to inform clinical care, public health interventions, infection prevention control, transmission, and clinical trials\nEfforts to develop a core clinical outcome set to maximize usability of data across a range of trials\nEfforts to determine adjunctive and supportive interventions that can improve the clinical outcomes of infected patients (e.g. steroids, high flow oxygen)\n\"\"\"\n\ntask_8 = \"\"\"What has been published concerning ethical considerations for research? What has been published concerning social sciences at the outbreak response?\nEfforts to articulate and translate existing ethical principles and standards to salient issues in COVID-2019\nEfforts to embed ethics across all thematic areas, engage with novel ethical issues that arise and coordinate to minimize duplication of oversight\nEfforts to support sustained education, access, and capacity building in the area of ethics\nEfforts to establish a team at WHO that will be integrated within multidisciplinary research and operational platforms and that will connect with existing and expanded global networks of social sciences.\nEfforts to develop qualitative assessment frameworks to systematically collect information related to local barriers and enablers for the uptake and adherence to public health measures for prevention and control. This includes the rapid identification of the secondary impacts of these measures. (e.g. use of surgical masks, modification of health seeking behaviors for SRH, school closures)\nEfforts to identify how the burden of responding to the outbreak and implementing public health measures affects the physical and psychological health of those providing care for Covid-19 patients and identify the immediate needs that must be addressed.\nEfforts to identify the underlying drivers of fear, anxiety and stigma that fuel misinformation and rumor, particularly through social media.\n\"\"\"\n\ntask_9 = \"\"\"What has been published about information sharing and inter-sectoral collaboration? What has been published about data standards and nomenclature? What has been published about governmental public health? What do we know about risk communication? What has been published about communicating with high-risk populations? What has been published to clarify community measures? What has been published about equity considerations and problems of inequity?\nMethods for coordinating data-gathering with standardized nomenclature.\nSharing response information among planners, providers, and others.\nUnderstanding and mitigating barriers to information-sharing.\nHow to recruit, support, and coordinate local (non-Federal) expertise and capacity relevant to public health emergency response (public, private, commercial and non-profit, including academic).\nIntegration of federal/state/local public health surveillance systems.\nValue of investments in baseline public health response infrastructure preparedness\nModes of communicating with target high-risk populations (elderly, health care workers).\nRisk communication and guidelines that are easy to understand and follow (include targeting at risk populations’ families too).\nCommunication that indicates potential risk of disease to all population groups.\nMisunderstanding around containment and mitigation.\nAction plan to mitigate gaps and problems of inequity in the Nation’s public health capability, capacity, and funding to ensure all citizens in need are supported and can access information, surveillance, and treatment.\nMeasures to reach marginalized and disadvantaged populations.\nData systems and research priorities and agendas incorporate attention to the needs and circumstances of disadvantaged populations and underrepresented minorities.\nMitigating threats to incarcerated people from COVID-19, assuring access to information, prevention, diagnosis, and treatment.\nUnderstanding coverage policies (barriers and opportunities) related to testing, treatment, and care\n\"\"\"\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"list_of_tasks = [task_1, task_2, task_3, task_4, task_5, task_6, task_7, task_8, task_9]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"def get_doc_vector(doc):\n    tokens = gensim.parsing.preprocess_string(doc)\n    vector = model.infer_vector(tokens)\n    return vector","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":" We get the vectors for the whole dataframe and the bullet points"},{"metadata":{"trusted":true},"cell_type":"code","source":"abstract_vectors = model.docvecs.vectors_docs\narray_of_tasks = [get_doc_vector(task) for task in list_of_tasks]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"train_df['abstract_vector'] = [vec for vec in abstract_vectors]","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"<a id='search'></a>\n\n## 4. Nearest Neigbors search\n"},{"metadata":{},"cell_type":"markdown","source":"Select only abstracts with at least 30 words"},{"metadata":{"trusted":true},"cell_type":"code","source":"train_df = train_df[train_df['abstract'].apply(lambda x: len(re.findall(r\"(?i)\\b[a-z]+\\b\", x))) > 40]\ntrain_df.shape\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"ball tree for Nearest Neighbor search fitted on the obtained abstract vectors"},{"metadata":{"trusted":true},"cell_type":"code","source":"train_array = train_df['abstract_vector'].values.tolist()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"from sklearn.neighbors import NearestNeighbors\nball_tree = NearestNeighbors(algorithm='ball_tree', leaf_size=20).fit(train_array)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"Query for the k nearest neighbours for each task :"},{"metadata":{"trusted":true},"cell_type":"code","source":"distances, indices = ball_tree.kneighbors(array_of_tasks, n_neighbors=3)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### Results\nPrint out texts abstracts and titles by categories "},{"metadata":{"trusted":true},"cell_type":"code","source":"for i, info in enumerate(list_of_tasks):\n    print(\"=\"*80, f\"\\n\\nTask = {info[:100]}\\n\", )\n    df =  train_df.iloc[indices[i]]\n    abstracts = df['abstract']\n    titles = df['title']\n    dist = distances[i]\n    for l in range(len(dist)):\n        print(f\" Text index = {indices[i][l]} \\n Distance = {distances[i][l]} \\n Title: {titles.iloc[l]} \\n Abstract extract: {abstracts.iloc[l][:200]}\\n\\n\")\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"<a id='clustering'></a>\n\n## 5. Clustering and visualisation "},{"metadata":{},"cell_type":"markdown","source":"Using our trained vectors, we can also use clustering and PCA to visualise the data and identify clusters of articles and locate the tasks regarding to those clusters"},{"metadata":{"trusted":true},"cell_type":"code","source":"\nfrom sklearn.cluster import KMeans\nfrom yellowbrick.cluster import KElbowVisualizer\n\nabstract_vectors = model.docvecs.vectors_docs\nkmeans = KMeans(init='k-means++', max_iter=300, random_state=42) \nvisualizer = KElbowVisualizer(kmeans, k=(2, 16))\nvisualizer.fit(abstract_vectors)\nvisualizer.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"This Elbow curves represents the \"clustering score\" for various values of K (it did effectively train the clustering for all the values) and automatically detects where the 'elbow\" is. The elbow is not really visible here but the Visualizer helps us detect it. \n\nI was expecting to find 4 as an optimal number of clusters because there are 4 different data sources. If we find one more it means that this clusterings does bring in some more information which is great ! so lets train on 6 clusters"},{"metadata":{"trusted":true},"cell_type":"code","source":"kmeans = KMeans(n_clusters=6, init='k-means++', max_iter=100, random_state=42) \nlabels = kmeans.fit_predict(train_array)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"Fit PCA then UMAP and get the vectors for the abstracts and the task :"},{"metadata":{"trusted":true,"_kg_hide-output":true},"cell_type":"code","source":"from sklearn.decomposition import PCA\nimport umap\npca = PCA(n_components=10).fit(train_array)\nabstract_pca = pca.transform(train_array)\ntasks_pca = pca.transform(array_of_tasks)\n\nmapper = umap.UMAP(n_components=3).fit(abstract_pca)\nabstract_umap = mapper.transform(abstract_pca)\ntasks_umap = mapper.transform(tasks_pca)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"train_df['abstract_pca_X'] = [a[0] for a in abstract_pca]\ntrain_df['abstract_pca_Y'] = [a[1] for a in abstract_pca]\ntrain_df['abstract_pca_Z'] = [a[2] for a in abstract_pca]\n\ntrain_df['abstract_umap_X'] = [a[0] for a in abstract_umap]\ntrain_df['abstract_umap_Y'] = [a[1] for a in abstract_umap]\ntrain_df['abstract_umap_Z'] = [a[2] for a in abstract_umap]\n\ntrain_df['labels'] = labels\ntrain_df['labels'] = train_df['labels'].astype(str)\n\ntasks_pca_X = [a[0] for a in tasks_pca]\ntasks_pca_Y = [a[1] for a in tasks_pca]\ntasks_pca_Z = [a[2] for a in tasks_pca]\n\n\ntasks_umap_X = [a[0] for a in tasks_umap]\ntasks_umap_Y = [a[1] for a in tasks_umap]\ntasks_umap_Z = [a[2] for a in tasks_umap]","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"Plot for visualisation"},{"metadata":{"trusted":true},"cell_type":"code","source":"import plotly.express as px\nfig = px.scatter_3d(train_df.sample(frac=0.1),  # plot only a fraction of the data \n                    x='abstract_umap_X', \n                    y='abstract_umap_Y', \n                    z='abstract_umap_Z',\n                    color='labels', \n                    symbol='labels', \n                    size_max=10, \n                    hover_data=['title'], \n                    height=800, \n                    width=1000)\n\nfig.add_scatter3d(x=tasks_umap_X, \n                  y=tasks_umap_Y, \n                  z=tasks_umap_Z, \n                  mode='markers', \n                  marker=dict(color='black', size=16), \n                  hovertext=[f'Task {i}' for i in range(1,10)],\n                  name='Tasks'\n                 )\nfig.update_layout(\n    title=\"UMAP Projection of abstracts\")\n\nfig.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"It seems that some of the tasks are regrouped in the same cluster which is interesting for two reasons : \n- we can group tasks together\n- we can assign clusters of articles to those groups of tasks\n\n* Task 1 = What is known about transmission, incubation, and environmental stability? \n* Task 2 = What do we know about COVID-19 risk factors? What have we learned from epidemiological studies?\n* Task 3 : What do we know about virus genetics, origin, and evolution?\n* Task 4 = What do we know about vaccines and therapeutics? What has been published concerning research and development and evaluation efforts of vaccines and therapeutics?\n* Task 5 : What do we know about the effectiveness of non-pharmaceutical interventions?\n* Task 6 : What do we know about diagnostics and surveillance? \n* Task 7: What has been published about medical care? \n* Task 8: What has been published concerning ethical considerations for research?\n* Task 9 : What has been published about information sharing and inter-sectoral collaboration? \n\nIt seems that there's a cluster that takes in the majority of the tasks.\n\n\nTasks 3 and 4 are together in a different cluster and task 2 is alone between some clusters. More investigation on cluster content could help us understand why this happens\n\n"},{"metadata":{"trusted":true},"cell_type":"code","source":"","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"name":"python","version":"3.6.4","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat":4,"nbformat_minor":4}