{"cells":[{"metadata":{},"cell_type":"markdown","source":"# TF-IDF Based Information Retrieval System with Summarization\n** Approach:**\n   > 1. Preprocessing of dataset and saving result to csv.\n   > 2. Compute cosine similarity of tf-idf vectors between query and dataset items, and return most relevant.\n   > 3. Display query, search results, and summaries as html elements.\n   > 4. Test our system for each subtask in each of the 10 tasks.\n   > 5. Visualize system performance for any given query through plotting the similarity scores of each of the retrieved documents with the query.\n    \n** Pros & Cons:**\n\n- **Pros:**\n> 1. IR through tf-idf is deterministic, requires no training time, and is very fast.\n> 2. Providing summaries helps user identify at a glance whether or not they're interested in a specific document.\n> 3. Built system is not task specific: works on all 10 tasks/subtasks and for any query. \n \n- **Cons:**\n> 1. System returns top n documents related to query, where n is a fixed user specified value that is not related to the query at hand. i.e: some queries might have more highly relevant documents than n, and some might have less than n and therefore return some documents which aren't very relevant.\n> 2. Summarization is done to each document disregarding the context of the query. This might lead to the summary excluding sentences that are important to the query. \n"},{"metadata":{},"cell_type":"markdown","source":"# Table of Contents:\n* [Util functions & Preprocessing](#1)\n* [TF-IDF & Cosine Similarity](#2)\n* [Summarization](#3)\n* [Testing All Tasks](#4)\n* [Evaluating System Performance](#5)\n"},{"metadata":{},"cell_type":"markdown","source":"<a id=\"1\"></a>\n# Util functions & Preprocessing\n  In this section we perform basic preprocessing of the dataset, removing any articles that don't contain our covid 19 terms, removing duplicates, and saving the result to csv."},{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true},"cell_type":"code","source":"import numpy as np \nimport pandas as pd\nimport re\nimport os\nimport json\nimport pickle\nfrom nltk.corpus import stopwords\nfrom collections import Counter\nfrom nltk import sent_tokenize, word_tokenize\n\n","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"d629ff2d2480ee46fbb7e2d37f6b5fab8052498a","_cell_guid":"79c7e3d0-c299-4dcb-8224-4455121ee9b0","trusted":true},"cell_type":"code","source":"# path to corpus of data\nCORONAVIRUS_LIBRARY_PATH = '../input/CORD-19-research-challenge'\n# path from which created csv is read\nRELEVANT_PAPERS_PATH = '../input/covid19csv/relevant_papers.csv'","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# checks if any of the keywords appear in text\ndef is_match(text, keywords):\n\n    return any(f'{keyterm} ' in text.lower() for keyterm in keywords)\n# loading files from json\ndef load_json(filename):\n    '''Load json file.'''\n    with open(filename, 'r') as f:\n        return json.load(f)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"ARTIFACTS = r'https?:\\/\\/.[^\\s\\\\]*|doi: medRxiv|preprint|\\[\\d+\\]|\\[\\d+\\, \\d+\\]'\n\n# retrieved from get_stopwords_from_corpus(sample) call in later cell\nCORPUS_STOPWORDS = ['as', 'of', 'during', 'the', 'disease', 'in', 'have', 'been', 'new', 'with',\n 'number', 'cases', 'is', 'or', 'not', 'to', 'it', 'on', 'its', 'outbreak', 'by',\n 'data', 'are', 'this', 'virus', 'and', 'epidemic', 'time', 'for', 'case', 'a', 'between', 'but',\n 'these', 'some', 'no', 'different', 'viral', 'transmission', 'clinical', 'from', 'we', 'confirmed',\n 'patients', 'were', 'study', 'china', 'infection', 'such', 'that', 'their', 'each', 'other', 'when',\n 'also', 'which', 'should', 'will', 'be', 'can', 'at', 'days', 'may', 'health', 'has', 'reported', 'an',\n 'infected', 'risk', 'was', '1', 'first', 'most', '.', 'used', 'two', 'using', ',', 'our', 'found', 'who', 'however,',\n 'more', 'they', 'all', '2', 'both', 'than', 'based', '', 'there', 'sars-cov-2', 'after', 'severe', 'respiratory', '3',\n 'one', 'including', 'et', 'use', 'control', 'had', 'if', 'only', 'into', 'al.,', 'made', 'under', 'could', 'cells',\n 'high', 'human', 'international', 'treatment', 'rate', 'without', '(', '=', 'patient', 'cell', 'results', 'available',\n 'model', 'figure', 'protein', 'display', 'copyright', 'granted', 'y', 'la', 'de', 'el', 'en', 'que', '(which', '•',\n 'holder', 'peer-reviewed)', 'license', 'author/funder,', 'medrxiv', 'perpetuity.']\n\n\n# credit: https://www.kaggle.com/rismakov/research-search-tool-and-article-summary/\nCOVID_19_TERMS = ['covid-19', 'covid 19','covid-2019','2019 novel coronavirus', 'corona virus disease 2019','coronavirus disease 19',\n    'coronavirus 2019','2019-ncov','ncov-2019',  'wuhan virus','wuhan coronavirus','wuhan pneumonia','NCIP','sars-cov-2','sars-cov2']\nVIRUS_TERMS = ['epidemic', 'pandemic', 'viral','virus','viruses','coronavirus', 'respiratory','infectious'] + COVID_19_TERMS","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"#this function saves all the data from the data set into a dataframe in the following format ['title' , 'text']\ndef get_all_data_without_duplicates():\n    '''Get all research paper data in dataset.\n\n    Returns\n    -------\n    pandas.DataFrame\n        Includes columns ['title' , 'text'].\n    '''\n    articles = []\n    for dirname, _, filenames in os.walk(CORONAVIRUS_LIBRARY_PATH):\n        for filename in filenames:\n            full_path = os.path.join(dirname, filename)\n            \n            if full_path.endswith('.json') :\n                article =load_json(full_path)\n                abstract=''\n                body=''\n                if(article.get(\"abstract\") != None):\n                    abstract = re.sub(ARTIFACTS, '', ' '.join(x['text'] for x in article['abstract']) , flags=re.MULTILINE).lower() \n                if(article.get(\"body_text\") != None):\n                    body =  re.sub(ARTIFACTS, '', ' '.join(x['text'] for x in article['body_text']) , flags=re.MULTILINE).lower() \n                authors = article['metadata']['authors']\n                list_authors =[]\n                for author in authors:\n                    if(len(author['middle'])==0):\n                        middle =\"\"\n                    else :\n                        middle = author['middle'][0]\n                    _authors =author['first']+ \" \"+ middle +\" \"+ author['last']\n\n                articles.append(\n                    [\n                        article['metadata']['title'], \n                        _authors,\n                        abstract+body,\n                    ]\n                )\n    return pd.DataFrame(articles, columns=['title', 'authors', 'text']).drop_duplicates()\n\n\n#this function filters the dataframe where its text contains COVID_19_TERMS\ndef filter_covid19_articles(df):\n    return df[\n        df['text'].apply(lambda x: is_match(x, set(COVID_19_TERMS)))\n    ]\n\n# this function gets stop words from the corpus , words repeated more than 5000 times\ndef get_stopwords_from_corpus(corpus_data_frame):\n    #this contains all the words in the corpus\n    sample = corpus_data_frame.text.str.cat(sep=' ').split(' ')\n    word_counts = Counter(sample)\n    max_word_count = max(word_counts.values())\n    threshold = 5000\n    return [word for word, count in word_counts.items() if count > threshold]\n\n# this function adds all stop words together \ndef get_all_stopwords():\n    return set(VIRUS_TERMS + CORPUS_STOPWORDS + stopwords.words('english'))\n\n# this function tokenize the sentece and only add words which are not in the stopwords list\ndef tokenize_sentences(data_frame_text):\n    return [\n        [\n            word.lower() for word in word_tokenize(sent) \n            if word not in get_all_stopwords()\n        ] for sent in sent_tokenize(data_frame_text)\n    ]","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"![](http://)Run the following cell **once only** to create the corpus csv (it should be then uploaded to input folder in a folder named \"covid19csv\" to be accessed later) "},{"metadata":{"trusted":true},"cell_type":"code","source":"#RUN ONCE TO GET THE RELEVANT DATA AND STORE IT TO CSV\n\n# sample = get_all_data_without_duplicates() # complete corpus\n# sample = filter_covid19_articles(sample) # corpus with relvant data to covid19\n# sample.to_csv('relevant_corpus.csv',index=False)\n\n# Run once to get stop words\n\n# CORPUS_STOPWORDS = get_stopwords_from_corpus(sample)\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"<a id=\"2\"></a>\n\n# TF-IDF\nIn this section we compute tf-idf vectors for both query and each of the dataset documents, then we use cosine similarity to obtain nearest n documents to the query. "},{"metadata":{"trusted":true},"cell_type":"code","source":"from sklearn.feature_extraction.text import TfidfVectorizer\nfrom sklearn.metrics.pairwise import linear_kernel\n# computes tf idf similarity between query and documents in dataset with n being max number of retrieved documents\ndef compute_tf_idf_similarity(query, docs, n):\n    tfidf_vectorizer = TfidfVectorizer()\n    tfidf = tfidf_vectorizer.fit_transform(docs)\n    query = tfidf_vectorizer.transform([query])\n    \n    cosine_similarities = linear_kernel(query, tfidf).flatten()\n#     related_docs_indices = cosine_similarities.argsort()[:-5:-1]\n    related_docs_indices = (-cosine_similarities).argsort()[:n]\n    \n    res = []\n    for i in range(0, len(related_docs_indices)):\n        res.append(docs[related_docs_indices[i]])\n        \n    res_df = pd.DataFrame(res, columns=['text'])\n    titles = []\n    authors = []\n    for j in range(0, len(res_df)):\n        for i in range(0, len(relevant_corpus)):\n            if(relevant_corpus['text'][i] == res_df['text'][j]):\n                titles.append(relevant_corpus['title'][i])\n                authors.append(relevant_corpus['authors'][i])\n                \n    res_df['title'] = titles\n    res_df['authors'] = authors\n    res_df = res_df[[\"title\",\"authors\", \"text\"]]\n    return res_df","execution_count":null,"outputs":[]},{"metadata":{"trusted":true,"collapsed":true},"cell_type":"code","source":"#sample call to test retrieving documents for a query\npd.set_option('max_colwidth', 500)\nrelevant_corpus = pd.read_csv(RELEVANT_PAPERS_PATH).loc[:]\ndf = pd.DataFrame(relevant_corpus)\ndocs = df['text'].tolist()\nquery = 'Integration of federal/state/local public health surveillance systems. '\nresults = compute_tf_idf_similarity(query, docs,5)\nresults\n    ","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"<a id=\"3\"></a>\n# Summarizing results\nIn this section we use Gensim's summarization tool to obtain summaries for a set of documents"},{"metadata":{"trusted":true},"cell_type":"code","source":"from gensim.summarization.summarizer import summarize\n# returns an array of strings, each element corresponding to a result document's summary\ndef summarize_results(results):\n    summarized_res = []\n    for i in range(0, len(results)):\n        ranked_sentences = summarize(results['text'][i], split=True, word_count=150) \n        summarized_res.append(ranked_sentences)\n        \n    string_summaries = []\n    for i in range(0, len(summarized_res)):\n        summary = \"\"\n        for j in range(0, len(summarized_res[i])):\n            summary += summarized_res[i][j]\n        string_summaries.append(summary)\n    return string_summaries","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"<a id=\"3\"></a>\n# Display Function\nHTML-formatted display to better visualize the system's operation"},{"metadata":{"trusted":true},"cell_type":"code","source":"from IPython.core.display import display, HTML\nimport math\n# html display of query, results and their respective summaries\ndef display_output(query, results, string_summaries):\n    display(HTML(f' <div style=\" padding: 25px\" > <h3 > Query: </h3> </br> <h4> {query} </h4> </div>  '))\n    display(HTML(f' <div style=\" padding: 25px\" > <h3 > Results: </h3> </div>'))\n    for i in range(0, len(results)):\n        title = results['title'][i]\n        authors = results['authors'][i]\n        if(pd.isna(title)):\n            title = \"No Title\"\n        summary = string_summaries[i]\n        display(HTML(f' <div style=\" padding: 25px\" >  <div> Title: <h5 style=\" display: inline\"> {title} </h5> </div><div> Author(s): <h5 style=\" display: inline\"> {authors} </h5> </div> <p> <b> Summary: </b> {summary}</p>  </div>  '))","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"<a id=\"4\"></a>\n# Testing Queries\nIn each cell you'll find a list of all the queries representing the subtasks in each task. To see results for a certain query, simply uncomment the line containing this query and make sure other queries are commented. "},{"metadata":{},"cell_type":"markdown","source":"* **Task 1:**"},{"metadata":{"trusted":true},"cell_type":"code","source":"relevant_corpus = pd.read_csv(RELEVANT_PAPERS_PATH).loc[:]\ndf = pd.DataFrame(relevant_corpus)\ndocs = df['text'].tolist()\n\n# queries, uncomment desired query and comment rest to see query results\n\nquery = 'Integration of federal/state/local public health surveillance systems. '\n# query = 'Range of incubation periods for the disease in humans (and how this varies across age and health status) and how long individuals are contagious, even after recovery.'\n# query = 'Prevalence of asymptomatic shedding and transmission (e.g., particularly children).'\n# query = 'Seasonality of transmission.'\n# query = 'Physical science of the coronavirus (e.g., charge distribution, adhesion to hydrophilic/phobic surfaces, environmental survival to inform decontamination efforts for affected areas and provide information about viral shedding).'\n# query = 'Persistence and stability on a multitude of substrates and sources (e.g., nasal discharge, sputum, urine, fecal matter, blood).'\n# query = 'Persistence of virus on surfaces of different materials (e,g., copper, stainless steel, plastic).'\n# query = 'Natural history of the virus and shedding of it from an infected person'\n# query = 'Implementation of diagnostics and products to improve clinical processes'\n# query = 'Disease models, including animal models for infection, disease and transmission'\n# query = 'Tools and studies to monitor phenotypic change and potential adaptation of the virus'\n# query = 'Immune response and immunity'\n# query = 'Effectiveness of movement control strategies to prevent secondary transmission in health care and community settings'\n# query = 'Effectiveness of personal protective equipment (PPE) and its usefulness to reduce risk of transmission in health care and community settings'\n# query = 'Role of the environment in transmission'\nresults = compute_tf_idf_similarity(query, docs,5)\nsummaries = summarize_results(results)\ndisplay_output(query, results,summaries)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* **Task 2**"},{"metadata":{"trusted":true},"cell_type":"code","source":"relevant_corpus = pd.read_csv(RELEVANT_PAPERS_PATH).loc[:]\ndf = pd.DataFrame(relevant_corpus)\ndocs = df['text'].tolist()\n\n# queries, uncomment desired query and comment rest to see query results\n\nquery = 'Data on potential risks factors'\n# query = 'Smoking, pre-existing pulmonary disease'\n# query = 'Co-infections (determine whether co-existing respiratory/viral infections make the virus more transmissible or virulent) and other co-morbidities'\n# query = 'Neonates and pregnant women'\n# query = 'Socio-economic and behavioral factors to understand the economic impact of the virus and whether there were differences.'\n# query = 'Transmission dynamics of the virus, including the basic reproductive number, incubation period, serial interval, modes of transmission and environmental factors'\n# query = 'Severity of disease, including risk of fatality among symptomatic hospitalized patients, and high-risk patient groups'\n# query = 'Susceptibility of populations'\n# query = 'Public health mitigation measures that could be effective for control'\nresults = compute_tf_idf_similarity(query, docs,5)\nsummaries = summarize_results(results)\ndisplay_output(query, results,summaries)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* **Task 3**"},{"metadata":{"trusted":true},"cell_type":"code","source":"relevant_corpus = pd.read_csv(RELEVANT_PAPERS_PATH).loc[:]\ndf = pd.DataFrame(relevant_corpus)\ndocs = df['text'].tolist()\n\n# queries, uncomment desired query and comment rest to see query results\n\nquery = 'Real-time tracking of whole genomes and a mechanism for coordinating the rapid dissemination of that information to inform the development of diagnostics and therapeutics and to track variations of the virus over time.'\n# query = 'Access to geographic and temporal diverse sample sets to understand geographic distribution and genomic differences, and determine whether there is more than one strain in circulation. Multi-lateral agreements such as the Nagoya Protocol could be leveraged.'\n# query = 'Evidence that livestock could be infected (e.g., field surveillance, genetic sequencing, receptor binding) and serve as a reservoir after the epidemic appears to be over.'\n# query = 'Evidence of whether farmers are infected, and whether farmers could have played a role in the origin.'\n# query = 'Surveillance of mixed wildlife- livestock farms for SARS-CoV-2 and other coronaviruses in Southeast Asia.'\n# query = 'Experimental infections to test host range for this pathogen.'\n# query = 'Animal host(s) and any evidence of continued spill-over to humans'\n# query = 'Socioeconomic and behavioral risk factors for this spill-over'\n# query = 'Sustainable risk reduction strategies'\nresults = compute_tf_idf_similarity(query, docs,5)\nsummaries = summarize_results(results)\ndisplay_output(query, results,summaries)\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* **Task 4**"},{"metadata":{"trusted":true},"cell_type":"code","source":"relevant_corpus = pd.read_csv(RELEVANT_PAPERS_PATH).loc[:]\ndf = pd.DataFrame(relevant_corpus)\ndocs = df['text'].tolist()\n\n# queries, uncomment desired query and comment rest to see query results\n\nquery = 'Effectiveness of drugs being developed and tried to treat COVID-19 patients.'\n# query = 'Clinical and bench trials to investigate less common viral inhibitors against COVID-19 such as naproxen, clarithromycin, and minocyclinethat that may exert effects on viral replication.'\n# query = 'Methods evaluating potential complication of Antibody-Dependent Enhancement (ADE) in vaccine recipients.'\n# query = 'Exploration of use of best animal models and their predictive value for a human vaccine.'\n# query = 'Capabilities to discover a therapeutic (not vaccine) for the disease, and clinical effectiveness studies to discover therapeutics, to include antiviral agents.'\n# query = 'Alternative models to aid decision makers in determining how to prioritize and distribute scarce, newly proven therapeutics as production ramps up. This could include identifying approaches for expanding production capacity to ensure equitable and timely distribution to populations in need.'\n# query = 'Efforts targeted at a universal coronavirus vaccine.'\n# query = 'Efforts to develop animal models and standardize challenge studies'\n# query = 'Efforts to develop prophylaxis clinical studies and prioritize in healthcare workers'\n# query = 'Approaches to evaluate risk for enhanced disease after vaccination'\n# query = 'Assays to evaluate vaccine immune response and process development for vaccines, alongside suitable animal models [in conjunction with therapeutics]'\nresults = compute_tf_idf_similarity(query, docs,5)\nsummaries = summarize_results(results)\ndisplay_output(query, results,summaries)\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* **Task 5**"},{"metadata":{"trusted":true},"cell_type":"code","source":"relevant_corpus = pd.read_csv(RELEVANT_PAPERS_PATH).loc[:]\ndf = pd.DataFrame(relevant_corpus)\ndocs = df['text'].tolist()\n\n# queries, uncomment desired query and comment rest to see query results\n\nquery = 'Resources to support skilled nursing facilities and long term care facilities.'\n# query = 'Mobilization of surge medical staff to address shortages in overwhelmed communities'\n# query = 'Age-adjusted mortality data for Acute Respiratory Distress Syndrome (ARDS) with/without other organ failure – particularly for viral etiologies'\n# query = 'Extracorporeal membrane oxygenation (ECMO) outcomes data of COVID-19 patients'\n# query = 'Outcomes data for COVID-19 after mechanical ventilation adjusted for age.'\n# query = 'Knowledge of the frequency, manifestations, and course of extrapulmonary manifestations of COVID-19, including, but not limited to, possible cardiomyopathy and cardiac arrest.'\n# query = 'Application of regulatory standards (e.g., EUA, CLIA) and ability to adapt care to crisis standards of care level.'\n# query = 'Approaches for encouraging and facilitating the production of elastomeric respirators, which can save thousands of N95 masks.'\n# query = 'Best telemedicine practices, barriers and faciitators, and specific actions to remove/expand them within and across state boundaries.'\n# query = 'Guidance on the simple things people can do at home to take care of sick people and manage disease.'\n# query = 'Oral medications that might potentially work.'\n# query = 'Use of AI in real-time health care delivery to evaluate interventions, risk factors, and outcomes in a way that could not be done manually.'\n# query = 'Best practices and critical challenges and innovative solutions and technologies in hospital flow and organization, workforce protection, workforce allocation, community-based support resources, payment, and supply chain management to enhance capacity, efficiency, and outcomes.'\n# query = 'Efforts to define the natural history of disease to inform clinical care, public health interventions, infection prevention control, transmission, and clinical trials'\n# query = 'Efforts to develop a core clinical outcome set to maximize usability of data across a range of trials'\n# query = 'Efforts to determine adjunctive and supportive interventions that can improve the clinical outcomes of infected patients (e.g. steroids, high flow oxygen)'\nresults = compute_tf_idf_similarity(query, docs,5)\nsummaries = summarize_results(results)\ndisplay_output(query, results,summaries)\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* **Task 6**"},{"metadata":{"trusted":true},"cell_type":"code","source":"relevant_corpus = pd.read_csv(RELEVANT_PAPERS_PATH).loc[:]\ndf = pd.DataFrame(relevant_corpus)\ndocs = df['text'].tolist()\n\n# queries, uncomment desired query and comment rest to see query results\n\nquery = 'Guidance on ways to scale up NPIs in a more coordinated way (e.g., establish funding, infrastructure and authorities to support real time, authoritative (qualified participants) collaboration with all states to gain consensus on consistent guidance and to mobilize resources to geographic areas where critical shortfalls are identified) to give us time to enhance our health care delivery system capacity to respond to an increase in cases'\n# query = 'Rapid design and execution of experiments to examine and compare NPIs currently being implemented. DHS Centers for Excellence could potentially be leveraged to conduct these experiments'\n# query = 'Rapid assessment of the likely efficacy of school closures, travel bans, bans on mass gatherings of various sizes, and other social distancing approaches'\n# query = 'Methods to control the spread in communities, barriers to compliance and how these vary among different populations'\n# query = 'Models of potential interventions to predict costs and benefits that take account of such factors as race, income, disability, age, geographic location, immigration status, housing status, employment status, and health insurance status'\n# query = 'Policy changes necessary to enable the compliance of individuals with limited resources and the underserved with NPIs'\n# query = 'Research on why people fail to comply with public health advice, even if they want to do so (e.g., social or financial costs may be too high)'\n# query = 'Research on the economic impact of this or any pandemic. This would include identifying policy and programmatic alternatives that lessen/mitigate risks to critical government services, food distribution and supplies, access to critical household supplies, and access to health diagnoses, treatment, and needed care, regardless of ability to pay'\nresults = compute_tf_idf_similarity(query, docs,5)\nsummaries = summarize_results(results)\ndisplay_output(query, results,summaries)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* **Task 7**"},{"metadata":{"trusted":true},"cell_type":"code","source":"relevant_corpus = pd.read_csv(RELEVANT_PAPERS_PATH).loc[:]\ndf = pd.DataFrame(relevant_corpus)\ndocs = df['text'].tolist()\n\n# queries, uncomment desired query and comment rest to see query results\n\nquery = 'Are there geographic variations in the rate of COVID-19 spread?'\n# query = 'Are there geographic variations in the mortality rate of COVID-19?'\n# query = 'Is there any evidence to suggest geographic based virus mutations?'\nresults = compute_tf_idf_similarity(query, docs,5)\nsummaries = summarize_results(results)\ndisplay_output(query, results,summaries)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* **Task 8**"},{"metadata":{"trusted":true},"cell_type":"code","source":"relevant_corpus = pd.read_csv(RELEVANT_PAPERS_PATH).loc[:]\ndf = pd.DataFrame(relevant_corpus)\ndocs = df['text'].tolist()\n\n# queries, uncomment desired query and comment rest to see query results\n\nquery = 'How widespread current exposure is to be able to make immediate policy recommendations on mitigation measures. Denominators for testing and a mechanism for rapidly sharing that information, including demographics, to the extent possible. Sampling methods to determine asymptomatic disease (e.g., use of serosurveys (such as convalescent samples) and early detection of disease (e.g., use of screening of neutralizing antibodies such as ELISAs)'\n# query = 'Efforts to increase capacity on existing diagnostic platforms and tap into existing surveillance platforms'\n# query = 'Recruitment, support, and coordination of local expertise and capacity (public, private—commercial, and non-profit, including academic), including legal, ethical, communications, and operational issues'\n# query = 'National guidance and guidelines about best practices to states (e.g., how states might leverage universities and private laboratories for testing purposes, communications to public health officials and the public)'\n# query = 'Development of a point-of-care test (like a rapid influenza test) and rapid bed-side tests, recognizing the tradeoffs between speed, accessibility, and accuracy'\n# query = 'Rapid design and execution of targeted surveillance experiments calling for all potential testers using PCR in a defined area to start testing and report to a specific entity. These experiments could aid in collecting longitudinal samples, which are critical to understanding the impact of ad hoc local interventions (which also need to be recorded)'\n# query = 'Separation of assay development issues from instruments, and the role of the private sector to help quickly migrate assays onto those devices'\n# query = 'Efforts to track the evolution of the virus (i.e., genetic drift or mutations) and avoid locking into specific reagents and surveillance/detection schemes'\n# query = 'Latency issues and when there is sufficient viral load to detect the pathogen, and understanding of what is needed in terms of biological and environmental sampling'\n# query = 'Use of diagnostics such as host response markers (e.g., cytokines) to detect early disease or predict severe disease progression, which would be important to understanding best clinical practice and efficacy of therapeutic interventions'\n# query = 'Policies and protocols for screening and testing'\n# query = 'Policies to mitigate the effects on supplies associated with mass testing, including swabs and reagents'\n# query = 'Technology roadmap for diagnostics'\n# query = 'Barriers to developing and scaling up new diagnostic tests (e.g., market forces), how future coalition and accelerator models (e.g., Coalition for Epidemic Preparedness Innovations) could provide critical funding for diagnostics, and opportunities for a streamlined regulatory environment'\n# query = 'New platforms and technology (e.g., CRISPR) to improve response times and employ more holistic approaches to COVID-19 and future diseases'\n# query = 'Coupling genomics and diagnostic testing on a large scale'\n# query = 'Enhance capabilities for rapid sequencing and bioinformatics to target regions of the genome that will allow specificity for a particular variant'\n# query = 'Enhance capacity (people, technology, data) for sequencing with advanced analytics for unknown pathogens, and explore capabilities for distinguishing naturally-occurring pathogens from intentional'\n# query = 'One Health surveillance of humans and potential sources of future spillover or ongoing exposure for this organism and future pathogens, including both evolutionary hosts (e.g., bats) and transmission hosts (e.g., heavily trafficked and farmed wildlife and domestic food and companion species), inclusive of environmental, demographic, and occupational risk factors'\nresults = compute_tf_idf_similarity(query, docs,5)\nsummaries = summarize_results(results)\ndisplay_output(query, results,summaries)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* **Task 9**"},{"metadata":{"trusted":true},"cell_type":"code","source":"relevant_corpus = pd.read_csv(RELEVANT_PAPERS_PATH).loc[:]\ndf = pd.DataFrame(relevant_corpus)\ndocs = df['text'].tolist()\n\n# queries, uncomment desired query and comment rest to see query results\n\nquery = 'Efforts to articulate and translate existing ethical principles and standards to salient issues in COVID-2019'\n# query = 'Efforts to embed ethics across all thematic areas, engage with novel ethical issues that arise and coordinate to minimize duplication of oversight'\n# query = 'Efforts to support sustained education, access, and capacity building in the area of ethics'\n# query = 'Efforts to establish a team at WHO that will be integrated within multidisciplinary research and operational platforms and that will connect with existing and expanded global networks of social sciences'\n# query = 'Efforts to develop qualitative assessment frameworks to systematically collect information related to local barriers and enablers for the uptake and adherence to public health measures for prevention and control. This includes the rapid identification of the secondary impacts of these measures. (e.g. use of surgical masks, modification of health seeking behaviors for SRH, school closures)'\n# query = 'Efforts to identify how the burden of responding to the outbreak and implementing public health measures affects the physical and psychological health of those providing care for Covid-19 patients and identify the immediate needs that must be addressed'\n# query = 'Efforts to identify the underlying drivers of fear, anxiety and stigma that fuel misinformation and rumor, particularly through social media'\nresults = compute_tf_idf_similarity(query, docs,5)\nsummaries = summarize_results(results)\ndisplay_output(query, results,summaries)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* **Task 10**"},{"metadata":{"trusted":true},"cell_type":"code","source":"relevant_corpus = pd.read_csv(RELEVANT_PAPERS_PATH).loc[:]\ndf = pd.DataFrame(relevant_corpus)\ndocs = df['text'].tolist()\n\n# queries, uncomment desired query and comment rest to see query results\n\nquery = 'Methods for coordinating data-gathering with standardized nomenclature'\n# query = 'Sharing response information among planners, providers, and others'\n# query = 'Understanding and mitigating barriers to information-sharing'\n# query = 'How to recruit, support, and coordinate local (non-Federal) expertise and capacity relevant to public health emergency response (public, private, commercial and non-profit, including academic)'\n# query = 'Integration of federal/state/local public health surveillance systems'\n# query = 'Value of investments in baseline public health response infrastructure preparedness'\n# query = 'Modes of communicating with target high-risk populations (elderly, health care workers)'\n# query = 'Risk communication and guidelines that are easy to understand and follow (include targeting at risk populations’ families too)'\n# query = 'Communication that indicates potential risk of disease to all population groups'\n# query = 'Misunderstanding around containment and mitigation'\n# query = 'Action plan to mitigate gaps and problems of inequity in the Nation’s public health capability, capacity, and funding to ensure all citizens in need are supported and can access information, surveillance, and treatment'\n# query = 'Measures to reach marginalized and disadvantaged populations'\n# query = 'Data systems and research priorities and agendas incorporate attention to the needs and circumstances of disadvantaged populations and underrepresented minorities'\n# query = 'Mitigating threats to incarcerated people from COVID-19, assuring access to information, prevention, diagnosis, and treatment'\n# query = 'Understanding coverage policies (barriers and opportunities) related to testing, treatment, and care'\nresults = compute_tf_idf_similarity(query, docs,5)\nsummaries = summarize_results(results)\ndisplay_output(query, results,summaries)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"<a id=\"5\"></a>\n# Evaluating Query Results\nIn this section, the relevance of the retrieved documents for a certain query can be visualized through our plot of cosine similarity per retrieved document "},{"metadata":{"trusted":true},"cell_type":"code","source":"import matplotlib.pyplot as plt\n# this returns the top n (max_documnets_number) cosine similarity scores\ndef get_top_cosine_similarity(query, docs, max_documnets_number):\n    tfidf_vectorizer = TfidfVectorizer()\n    tfidf = tfidf_vectorizer.fit_transform(docs)\n    query = tfidf_vectorizer.transform([query])\n    \n    cosine_similarities = linear_kernel(query, tfidf).flatten()\n    cosine_similarities_list = cosine_similarities.tolist()\n    cosine_similarities_list.sort(reverse=True)\n    top_cosine_similarity = cosine_similarities_list[0:max_documnets_number]\n    return top_cosine_similarity\n#this plots documents against cosine similarity scores\ndef plot_top_cosine_similarity(query, docs, max_documnets_number):\n    top_cosine_simlrity = get_top_cosine_similarity(query, docs, max_documnets_number)\n    result_number = np.arange(start=1, stop=max_documnets_number+1)\n    plt.figure(figsize=(8,6))\n    plt.bar(result_number, top_cosine_simlrity, align='center', alpha=0.5)\n    plt.xlabel('Document Number')\n    plt.ylabel('Cosine Simlarity')\n    plt.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"Write any query here to visualize the system's performance (measured in cosine similarity score) in regards to document relevance to query. "},{"metadata":{"trusted":true},"cell_type":"code","source":"query = 'Effectiveness of drugs being developed and tried to treat COVID-19 patients.'\nplot_top_cosine_similarity(query, docs, 5)","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}