{"cells":[{"metadata":{"_uuid":"72e8abcc5687c2bcbf891c52a38f013d1784a807"},"cell_type":"markdown","source":"This is a first draft of an analysis of the Genetic Variant Classification dataset. Comments and feedback are welcome!\n\n**Methodology:**\n - Initially the data is cleaned: NaN entries are either droped or imputed, and most numerical data is normalized to lie in the range [0,1]\n - Several new features are engineered, the most novel being a 'weight' feature for the CLNDN, MC and Consequence columns; see below for further details.\n - The dataset is then balanced (initially ~75% of data belongs to class 0) and categorical data is encoded numerically (only a selection of the columns are used in this first draft).\n - A RandomForest classifier is repeatedly applied to splits of the dataset (90% train, 10% test) and accuracy statistics are collected (see \"Findings\" below).\n\n**Findings:**\n\n   Following [this kernel](https://www.kaggle.com/vasilyb/clinvar-identifying-conflicting-genetic-variants), we use an (averaged) AUROC score for measuring the efficacy the RandomForestClassifier. Iterating over various samples of the processed data yields an average AUROC score greater than 0.8. By adjusting the RandomForestClassifier, it seems possible to obtain a TPR / FPR ratio of 0.8 / 0.2.\n   \n**ToDo List:**\n1. Incorporate other key features into the model, e.g.  cDNA / CDS / Protein_positions, Amion_Acids, Exon, Codons\n2. Hypertune the parameters for the RandomForestClassifier to search for more optimal prediction models\n3. Apply other common ML algorithms"},{"metadata":{"trusted":true,"_uuid":"bb1de942daf1c7bb762f5dd95345de338e1c7169"},"cell_type":"code","source":"import numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\nimport seaborn as sns\nfrom scipy.stats import binom\nimport matplotlib.pyplot as plt\nfrom sklearn.impute import SimpleImputer\nfrom sklearn.ensemble import RandomForestClassifier\nfrom sklearn.metrics import accuracy_score, roc_curve, auc, f1_score, recall_score, precision_score, classification_report\nfrom sklearn.model_selection import train_test_split\n\nverbose = False\n\n#alpha is a parameter to control how much of the final dataset is used in training the random forest classifier\nalpha = 0.9\n\n#n_models determines how many times to build and test a classifier from the final dataset\nn_models = 20\n\n#this is a random parameter to control the random state of the model, for reproducibility.\nrnd=0","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"f4d9251610cca065ae9a5f2db3eddabb21034d30"},"cell_type":"markdown","source":"The \"relative_weights\" function in the next code box is used to engineer several new features in the dataset. Here, the procedure is briefly described:\n\n        The columns CLNDN, MC, and Consequence of the given dataset are 'multi-entires', in the sense that each entry contains a string that lists (possibly several) sub-features of the particular entry in the data set. These sub-features are first collected into a dictionary each entry of which is of the form \n\n        \"sub-feature\" : [overall count of sub-frequency, count of sub-frequency in class 1]\n\n        Each sub-feature is then assigned a class, 0 or 1, based on whether rows with sub-feature were more often 0 or 1 (strategy is winner-take-all; the borderline case of equal frequency will not matter, see the expression for predictive power below). Sub-features are also assigned a weight, a measure of their predictive power, given by\n\n        rho = count of sub-feature in class 1 / overall count of sub-feature\n        predictive power = (2 rho - 1)^2\n\n        Thus, if the sub-feature appeared equally often in both classes, the predictive power is 0, while if it appeared more often in one class or the other, the predictive power is close to 1. Finally, we use the sub-feature class to weight each row using the formula\n\n        weight = (1 / # sub-features in this row) * Sum [(-1)^ (sub-feature class) * predictive power]\n\n        Here the sum is taken over all sub-features appearing in this row. The effect of this is that, if a row contains sub-features, the majority of which belong to class 0, then the row is assigned a weight near 1, while if the majority of the sub-features in a row belong to class 1, the row is assigned weight -1. Weights near 0 mean that the sub-features appearing in the given row are relatively balanced between the two classes.\n\nThe remaining helper functions in the next code box are just for convenience in pre-processing other parts of the dataset."},{"metadata":{"trusted":true,"_uuid":"63acadb4e66592c925fbf9de73a9286ff8db1678"},"cell_type":"code","source":"def relative_weights(df, class_name, col_name, delimiter, synonyms ={}, reduce = False):\n    #df is the input dataframe\n    #col_name is the name of the column whose relative weights we want to compute\n    #delimiter is how the entries in col_name are divided into classes\n    #synonyms lists any synonyms among words that may appear in col_name entries, e.g. not provided = not specified\n\n    counts = {}\n    for index, row in df.iterrows():\n        #extract the words for this row\n        new_words = row[col_name].split(delimiter)\n        #make synonym substitutions and remove any duplicates\n        new_words = list(set([synonyms.get(j,j) for j in new_words]))\n        for n in new_words:\n            #for existing words, we increase the overall count and the class frequency\n            #if the word is new, create an entry [1, (class of this row)]\n            if n in counts:\n                counts[n][0] +=1\n                counts[n][1] += row[class_name]\n            else:\n                counts[n] = [1,row[class_name]]\n    n_entries = df.shape[0]\n    weights = dict([(key, (np.power((2*counts[key][1]/counts[key][0] -1),2), np.floor(0.5 + counts[key][1]/counts[key][0]))) \n                      for key in counts])\n\n    #create a subfunction to get the weight of a given entry\n    #this will allow us to use a lambda function to map over rows of the dataframe\n    def get_weight(row, dictionary, delimiter_sub, syns):\n        kwds = row.split(delimiter_sub)\n        #make synonym substitutions\n        kwds = list(set([syns.get(j,j) for j in kwds]))\n        return np.sum([np.power((-1),dictionary[j][1]) *dictionary[j][0] for j in kwds])/len(kwds)\n        \n    return df[col_name].apply(lambda x: get_weight(x, weights, delimiter,synonyms))\n\n#for converting Chromosome numbers to strings\ndef to_int(s):\n    try:\n        return int(s)\n    except ValueError:\n        if s == 'X':\n            return 23\n        else:\n            return 24\n    \n#for shortening base-pair entries\ndef string_to_short(s):\n    if len(s)>1:\n        return 'L'\n    else:\n        return s\n    \n#Extract training and testing data from a dataset\ndef get_train_test(df, alpha = 0.9, clss=None):\n    train = df.sample(frac = 0.9, axis = 0)\n    test = df[~df.index.isin(train.index)]\n    \n    cols_not_class = [j for j in df.columns.values]\n    if clss in cols_not_class:\n        cols_not_class.remove(clss)\n    return train[cols_not_class], train[clss], test[cols_not_class], test[clss]","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"868401fc4518d52a37cfd2c9eca8c57d812921cd"},"cell_type":"markdown","source":"The next code box contains the initial pre-processing of the data. A few remarks are in order:\n1. We begin by dropping columns which have 50% of their entries as NaN; it seems difficult to use such sparse data in making predictions.\n2. Some entries in the CHROM column contain numeric entries, some alphabetic (e.g. 22 vs. '22'). All are converted to integers (and the extra entires, 'X' and 'MT' are assigned numbers 23 and 24 respectively.\n3. The columns REF, ALT, and Allele contain base-pair sequences. We add new features marking the lengths of the base-pairs appearing (as was done in [this kernel](https://www.kaggle.com/vasilyb/clinvar-identifying-conflicting-genetic-variants)), and we also modify entries in REF, ALT, and Allele with long base-pair sequences (length > 1) to an entry 'L'.\n4. LoFtool, ORIGIN, CADD_PHRED, CADD_RAW are numeric with missing values that we impute; for ORIGIN we use the median, since its entries are all integers, while for the other columns we impute using the mean.\n5. The data in POS seems to carry data about where, in a given chromosome, the mutation occurs. Since different chromosomes have different lengths, we normalize entries in the POS column to take values between 0 and 1, so that the scale of entries in POS will not bias the model.\n6. Data in the IMPACT column has categorical descriptions that seem to imply an order (HIGH - MEDIUM - LOW), so we convert these to evenly-spaced floats in [0,1]."},{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true,"scrolled":true},"cell_type":"code","source":"data = pd.read_csv(\"../input/clinvar_conflicting.csv\")\n\nprint(\"Processing data...\")\n\n#drop NA values up to 50%\ndata = data.dropna(axis=1, thresh=.5*data.shape[0])\n\n#convert entries in the Chromosome column to integers\ndata.CHROM = pd.Series(data.CHROM.map(lambda x: to_int(x)), index = data.CHROM.index)\n\n#drop rows with few NA features in certain columns\ndata = data.drop(data[ (data['STRAND'].isna()) | (data['BIOTYPE'].isna()) | (data['SYMBOL'].isna())].index, axis = 0)\n\n#'Feature Type' has a unique entry, so we drop this feature\ndata = data.drop(['Feature_type'], axis = 1)\n\n#start looking at single / multiple base entries. We see most entries are single bases\nsingle_base = ['T','A','G','C','-']\n\nI = data[~data.REF.isin(single_base)].index\nI=I.union(data[~data.ALT.isin(single_base)].index)\nI=I.union(data[~data.Allele.isin(single_base)].index)\n\n#We'll address this in two ways: we'll add a feature that counts the length of the bases involved in the REF, ALT, Allele category\n#We'll also condense the longer base sequences into a category 'L' -- this allows the distinctions between G,C,T,A,- to still appear in thse features    \ndata['REF_LEN'] = pd.Series(data['REF'].apply(lambda x: len(x)), index = data['REF'].index)\ndata['ALT_LEN'] = pd.Series(data['ALT'].apply(lambda x: len(x)), index = data['ALT'].index)\ndata['Allele_LEN'] = pd.Series(data['Allele'].apply(lambda x: len(x)), index = data['Allele'].index)\n    \ndata['REF'] = pd.Series(data['REF'].apply(string_to_short), data['REF'].index)\ndata['ALT'] = pd.Series(data['ALT'].apply(string_to_short), data['ALT'].index)\ndata['Allele'] = pd.Series(data['Allele'].apply(string_to_short), data['Allele'].index)\n\n#Here we start with some imputation\nimp_mean = SimpleImputer(strategy = 'mean')\nimp_med = SimpleImputer(strategy = 'median')\n\ndata['ORIGIN'] = pd.Series(imp_med.fit_transform(data.ORIGIN.values.reshape((-1,1))).flatten(), index = data.ORIGIN.index)\ndata['LoFtool'] = pd.Series(imp_mean.fit_transform(data.LoFtool.values.reshape((-1,1))).flatten(), index = data.LoFtool.index)\ndata['CADD_PHRED'] = pd.Series(imp_mean.fit_transform(data.CADD_PHRED.values.reshape((-1,1))).flatten(), index = data.CADD_PHRED.index)\ndata['CADD_RAW'] = pd.Series(imp_mean.fit_transform(data.CADD_RAW.values.reshape((-1,1))).flatten(), index = data.CADD_RAW.index)\n\n#Finally, we normalize numerical columns\ncols_to_normalize = ['AF_ESP', 'AF_EXAC', 'AF_TGP', 'LoFtool', 'CADD_PHRED', 'CADD_RAW']\nfor c in cols_to_normalize:\n    data[c] = pd.Series((lambda x: (x-x.min())/(x.max() - x.min()))(data[c]), index = data[c].index)\n\n#We also normalize the position factor\n#within each chromosome, we rescale the position to a numerical value between 0 and 1 representing its relative position in the chromosome\n#We'll use the largest position for a given chromosome as a proxy for its actual maximal length; this needs to be replaced with actual maximal lengths\nMaxes = [data[data['CHROM']==j].POS.max() for j in data.CHROM.unique()]\ndata.POS = data.apply(lambda x: x.POS / Maxes[x.CHROM-1], axis = 1)\n    \n#replace NaN entries in the Codons, Amino_Acids, and MC columns with a class 'unkn'.\ndata.Codons.fillna('unkn', inplace=True)\ndata.Amino_acids.fillna('unkn', inplace=True)\ndata.MC.fillna('unkn', inplace=True)\n\n#convert the impact scores to numerical values\nimpact_dict = {\"MODIFIER\":0, \"LOW\": 0.33333, \"MODERATE\": 0.66667, \"HIGH\": 1}\ndata[\"IMPACT\"] = data[\"IMPACT\"].apply(lambda x:impact_dict[x])\n    \n#create the weights column\ndata['CLNDN_WTS'] = relative_weights(data, \"CLASS\", \"CLNDN\", \"|\", {\"not_specified\" : \"not_provided\"})\ndata['MC_WTS'] = relative_weights(data, \"CLASS\", \"MC\", \",\")\ndata['Consequence_WTS'] = relative_weights(data, \"CLASS\", \"Consequence\", \"&\")\n    \n    \nprint(\"Finished!\")\n#Some summary info about data so far\nif verbose:\n    for c in data.columns:\n        print(\"Column: \", c, \"\\n\")\n        print(data[c].unique())\n        print(\"\\n\")\n    sns.countplot(x=\"CLASS\", data = data)","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"aa6a5a71e3356bd85e5f8f96b89bbf2e53c0a569"},"cell_type":"markdown","source":"We include a separate plot showing the result of the new feature engineered using the relative_weights function (the plots for MC_WTS and Consqeuence_WTS are essentially the same, so we plot them together)."},{"metadata":{"trusted":true,"_uuid":"bc790fdea8aa339cb07aec074dd2a9899c2caafb"},"cell_type":"code","source":"pd.Series(sorted(data.CLNDN_WTS.values, reverse = True)).plot()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true,"_uuid":"5bee8e5da011929abf9a003b9fff01343d2720ca"},"cell_type":"code","source":"pd.Series(sorted(data.MC_WTS.values, reverse = True)).plot()\npd.Series(sorted(data.Consequence_WTS.values, reverse = True)).plot()","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"063f2dbb4993deb8c73e5fd5025fd5911dd9e820"},"cell_type":"markdown","source":"The modeling occurs in the next code box. A few notes are in order:\n1. Only some of the columns are used. There are others that likely contain useful information; this is just a first draft and future updates will take these other columns into account.\n2. The data turns out to be fairly imbalanced (nearly 75% of entries are class 0); here the data is balanced by selecting a subset of class 0 entries containing as many elements as there are in the class 1 entries, so that the balanced dataset has "},{"metadata":{"trusted":true,"_uuid":"b11fab591f50447097b3ca70bd9715ad4bc413b4","scrolled":false},"cell_type":"code","source":"#select features for the model\nfeatures_columns = [\"CHROM\", \"POS\", \"REF\", \"ALT\", \"AF_ESP\", \"AF_EXAC\", \"AF_TGP\", \"CLNVC\", \"ORIGIN\", \"CLASS\", \"Allele\", \"IMPACT\", \"STRAND\",\n                   \"LoFtool\", \"CADD_PHRED\", \"CADD_RAW\", \"REF_LEN\", \"ALT_LEN\",\"Allele_LEN\", \"CLNDN_WTS\", \"MC_WTS\", \"Consequence_WTS\"]\n\n#here we extract a balanced dataset:\ndata_balanced = pd.concat([data[data['CLASS']==0].sample(n=data[data['CLASS']==1].shape[0], axis = 0), data[data['CLASS']==1]])\n\n#encode categorical data and extract train / test data\ndata_bal_encoded = pd.get_dummies(data_balanced[features_columns])\nfeatures = data_bal_encoded[data_bal_encoded.columns.drop(\"CLASS\")]\nlabels = data_bal_encoded.CLASS\n\n#model_parameters\na=[150, 15]\nclf = RandomForestClassifier(n_estimators = a[0], min_samples_split = a[1])\n\nplt.figure()\n\nAUCS = []\nf1s =[]\n\nfor i in range(n_models):\n    #split and fit the data\n    features_train, features_test, labels_train, labels_test = train_test_split(features, labels, train_size = alpha, test_size = 1-alpha)\n    roc_probs = clf.fit(features_train, labels_train).predict_proba(features_test)\n    y_pred = clf.predict(features_test)\n    \n    #feedback from this model\n    print(\"Model %i trained\"%(i+1))\n    if verbose: \n        print(classification_report(labels_test, y_pred))\n        print(\"Model %i AUROC score: %0.3f\"%((i+1), auc(fpr,tpr)))\n    \n    #capture data\n    fpr, tpr, threshs = roc_curve(labels_test.values, roc_probs[:,1])\n    plt.plot(fpr, tpr)\n    AUCS.append(auc(fpr,tpr))\n    \nplt.plot([0,1],[0,1], linestyle = '--')\nplt.xlabel(\"FPR\")\nplt.ylabel(\"TPR\")\nplt.suptitle(\"AUROC Curves\")\nplt.title(\"Average AUROC: %0.3f; Standard Deviation: %0.3f\"%(np.mean(AUCS), np.std(AUCS)))\nplt.show()","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"name":"python","version":"3.6.6","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat":4,"nbformat_minor":1}