{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"# What is about \n\nAdaptation of https://nbviewer.jupyter.org/github/KrishnaswamyLab/MAGIC/blob/master/python/tutorial_notebooks/bonemarrow_tutorial.ipynb\n\nSee also MAGIC main page: https://github.com/KrishnaswamyLab/MAGIC\n\nThere are certain differences in some version of notebook. \nE.g. data are taken directly from GEO - at there much more cells (Version 1). \n\nVersions:\n\n2 Use exactly the same data as original tutorial \n\n1 Use different version of the same dataset - directly obtained from GEO and strangely enough it contains more cells.\nData preparation notebook: https://www.kaggle.com/alexandervc/conversion-gse72857-txt-data-from-geo-to-h5ad\nSo the figures a bit different than in original tutorial\n\n\n## Python MAGIC Bone Marrow tutorial\n\nMAGIC imputes missing data values on sparse data sets, restoring the structure of the data\nIt also proves dimensionality reduction and gene expression visualizations\nMAGIC can be performed on a variety of datasets\nHere, we show the effectiveness of MAGIC on epithelial-to-mesenchymal transition (EMT) data\nMarkov Affinity-based Graph Imputation of Cells (MAGIC) is an algorithm for denoising and transcript recover of single cells applied to single-cell RNA sequencing data, as described in Van Dijk D et al. (2018), Recovering Gene Interactions from Single-Cell Data Using Data Diffusion, Cell https://www.cell.com/cell/abstract/S0092-8674(18)30724-4.\n\nThis tutorial shows loading, preprocessing, MAGIC imputation and visualization of myeloid and erythroid cells in mouse bone marrow, as described by Paul et al., 2015. You can edit it yourself at https://colab.research.google.com/github/KrishnaswamyLab/MAGIC/blob/master/python/tutorial_notebooks/bonemarrow_tutorial.ipynb\n\n## Table of Contents\n\nInstallation\n\nLoading data\n\nData preprocessing\n\nRunning MAGIC\n\nVisualizing gene-gene interactions\n\nVisualizing cell trajectories with PCA on MAGIC and PHATE\n\nUsing MAGIC data in downstream analysis\n","metadata":{}},{"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\n\n# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","execution":{"iopub.status.busy":"2021-07-31T18:01:31.054722Z","iopub.execute_input":"2021-07-31T18:01:31.055274Z","iopub.status.idle":"2021-07-31T18:01:31.066985Z","shell.execute_reply.started":"2021-07-31T18:01:31.055222Z","shell.execute_reply":"2021-07-31T18:01:31.066159Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"import time\nt0start = time.time()\nt0 = time.time()\n\n\n!pip install --user magic-impute\n\nimport magic\nimport scprep # scprep provides an all-in-one framework for loading, preprocessing, and plotting matrices in Python, with a focus on single-cell genomics.\n\n\nimport numpy as np\nimport pandas as pd\nimport matplotlib\nimport matplotlib.pyplot as plt\n\n# Matplotlib command for Jupyter notebooks only\n%matplotlib inline\n\n!pip install scanpy\nimport scanpy as sc\nimport anndata\n\nimport scipy \n\n\n\nimport pandas as pd\nimport numpy as np\nimport os\nimport sys\n\nprint(np.round(time.time()-t0,1) , 'seconds passed')\n","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:01:31.086976Z","iopub.execute_input":"2021-07-31T18:01:31.08737Z","iopub.status.idle":"2021-07-31T18:01:56.427869Z","shell.execute_reply.started":"2021-07-31T18:01:31.087328Z","shell.execute_reply":"2021-07-31T18:01:56.42688Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Loading data\n\nscprep can load several formats, but seems not h5ad, so we will use scanpy\n\nAbout scprep: Load your data using one of the following scprep.io methods: load_csv, load_tsv, load_fcs, load_mtx, load_10x. You can read about how to use them with help(scprep.io.load_csv) or on https://scprep.readthedocs.io/.\n","metadata":{}},{"cell_type":"code","source":"import time\nt0 = time.time()\n\nif 1:\n    fn = '/kaggle/input/scrnaseq-bone-marrow-collection-of-datasets/GSE72857_Paul_Amit_2015_CSVdataFromKrishnaswamyLab/BMMC_myeloid.csv'\n    bmmsc_data = scprep.io.load_csv(fn)# 'https://github.com/KrishnaswamyLab/PHATE/raw/master/data/BMMC_myeloid.csv.gz')\n\nelse:\n    fn = '/kaggle/input/scrnaseq-bone-marrow-collection-of-datasets/GSE72857_Mouse_BoneMarrow_Lin-cKitSca1-MyeloidProgenitors_Y2015_10368Cells_GEO_counts_Paul_Amit.h5ad'\n    adata = sc.read(fn)\n    adata_orig = adata.copy()\n    print(adata)\n    bmmsc_data = adata.to_df()\n    bmmsc_data.index = adata.obs['Well_ID']\n\nprint(np.round(time.time()-t0,1) , 'seconds passed')\n\nbmmsc_data\n","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:01:56.429375Z","iopub.execute_input":"2021-07-31T18:01:56.429699Z","iopub.status.idle":"2021-07-31T18:02:40.271012Z","shell.execute_reply.started":"2021-07-31T18:01:56.429667Z","shell.execute_reply":"2021-07-31T18:02:40.269722Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Data Preprocessing","metadata":{}},{"cell_type":"markdown","source":"## Filtering your data\n\nAfter loading your data, you're going to want to determine the molecule per cell and molecule per gene cutoffs with which to filter the data, in order to remove lowly expressed genes and cells with a small library size.","metadata":{}},{"cell_type":"code","source":"scprep.plot.plot_library_size(bmmsc_data, cutoff=1000)\n","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:02:40.273006Z","iopub.execute_input":"2021-07-31T18:02:40.273364Z","iopub.status.idle":"2021-07-31T18:02:41.020095Z","shell.execute_reply.started":"2021-07-31T18:02:40.273329Z","shell.execute_reply":"2021-07-31T18:02:41.019215Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"bmmsc_data = scprep.filter.filter_library_size(bmmsc_data, cutoff=1000)\nbmmsc_data.head()","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:02:41.023685Z","iopub.execute_input":"2021-07-31T18:02:41.023982Z","iopub.status.idle":"2021-07-31T18:02:41.568619Z","shell.execute_reply.started":"2021-07-31T18:02:41.023953Z","shell.execute_reply":"2021-07-31T18:02:41.567542Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"We should also remove genes that are not expressed above a certain threshold, since they are not adding anything valuable to our analysis.","metadata":{}},{"cell_type":"code","source":"bmmsc_data = scprep.filter.filter_rare_genes(bmmsc_data, min_cells=10)\nbmmsc_data.head()","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:02:41.569831Z","iopub.execute_input":"2021-07-31T18:02:41.570145Z","iopub.status.idle":"2021-07-31T18:02:41.898963Z","shell.execute_reply.started":"2021-07-31T18:02:41.570112Z","shell.execute_reply":"2021-07-31T18:02:41.897945Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Normalizing your data\nAfter filtering, the next steps are to perform library size normalization and transformation. Log transformation is frequently used for single-cell RNA-seq, however, this requires the addition of a pseudocount to avoid infinite values at zero. We instead use a square root transform, which has similar properties to the log transform but has no problem with zeroes.\n","metadata":{}},{"cell_type":"code","source":"t0 = time.time()\nbmmsc_data = scprep.normalize.library_size_normalize(bmmsc_data)\nbmmsc_data = scprep.transform.sqrt(bmmsc_data)\nprint(np.round(time.time()-t0,1) , 'seconds passed')\nbmmsc_data.head()","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:02:41.900112Z","iopub.execute_input":"2021-07-31T18:02:41.900392Z","iopub.status.idle":"2021-07-31T18:02:42.420719Z","shell.execute_reply.started":"2021-07-31T18:02:41.900364Z","shell.execute_reply":"2021-07-31T18:02:42.419872Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Running MAGIC\n\nNow that your data has been preprocessed, you are ready to run MAGIC.\n\n## Creating the MAGIC operator\n\nIf you don't specify any parameters, the following line creates an operator with the following default values: knn=5, decay=1, t=3.","metadata":{}},{"cell_type":"code","source":"magic_op = magic.MAGIC()\n","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:02:42.421917Z","iopub.execute_input":"2021-07-31T18:02:42.422254Z","iopub.status.idle":"2021-07-31T18:02:42.426417Z","shell.execute_reply.started":"2021-07-31T18:02:42.422224Z","shell.execute_reply":"2021-07-31T18:02:42.425271Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Running MAGIC with gene selection\n\nThe magic_op.fit_transform function takes the normalized data and an array of selected genes as its arguments. If no genes are provided, MAGIC will return a matrix of all genes. The same can be achieved by substituting the array of gene names with genes='all_genes'.","metadata":{}},{"cell_type":"code","source":"t0 = time.time()\nbmmsc_magic = magic_op.fit_transform(bmmsc_data, genes=[\"Mpo\", \"Klf1\", \"Ifitm1\"])\nprint(np.round(time.time()-t0,1) , 'seconds passed')\n\nbmmsc_magic.head()","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:02:42.428771Z","iopub.execute_input":"2021-07-31T18:02:42.429073Z","iopub.status.idle":"2021-07-31T18:02:48.387365Z","shell.execute_reply.started":"2021-07-31T18:02:42.429023Z","shell.execute_reply":"2021-07-31T18:02:48.386459Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Visualizing gene-gene relationships\n\nWe can see gene-gene relationships much more clearly after applying MAGIC. Note that the change in absolute values of gene expression is not meaningful - the relative difference is all that matters.","metadata":{}},{"cell_type":"code","source":"fig, (ax1, ax2) = plt.subplots(1,2, figsize=(16, 6))\n\nscprep.plot.scatter(x=bmmsc_data['Mpo'], y=bmmsc_data['Klf1'], c=bmmsc_data['Ifitm1'],  ax=ax1,\n                    xlabel='Mpo', ylabel='Klf1', legend_title=\"Ifitm1\", title='Before MAGIC')\n\nscprep.plot.scatter(x=bmmsc_magic['Mpo'], y=bmmsc_magic['Klf1'], c=bmmsc_magic['Ifitm1'], ax=ax2,\n                    xlabel='Mpo', ylabel='Klf1', legend_title=\"Ifitm1\", title='After MAGIC')\n\nplt.tight_layout()\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:02:48.388964Z","iopub.execute_input":"2021-07-31T18:02:48.389243Z","iopub.status.idle":"2021-07-31T18:02:49.211299Z","shell.execute_reply.started":"2021-07-31T18:02:48.389217Z","shell.execute_reply":"2021-07-31T18:02:49.210195Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"The original data suffers from dropout to the point that we cannot infer anything about the gene-gene relationships. As you can see, the gene-gene relationships are much clearer after MAGIC. These relationships also match the biological progression we expect to see - Ifitm1 is a stem cell marker, Klf1 is an erythroid marker, and Mpo is a myeloid marker.","metadata":{}},{"cell_type":"markdown","source":"## Setting the MAGIC operator parameters\n\nIf you wish to modify any parameters for your MAGIC operator, you change do so without having to recompute intermediate values using the magic_op.set_params method. Since our gene-gene relationship here appears a little too noisy, we can increase t a little from the default value of 3 up to a larger value like 5.","metadata":{}},{"cell_type":"code","source":"magic_op.set_params(t=5)\n","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:02:49.212832Z","iopub.execute_input":"2021-07-31T18:02:49.213233Z","iopub.status.idle":"2021-07-31T18:02:49.221567Z","shell.execute_reply.started":"2021-07-31T18:02:49.213184Z","shell.execute_reply":"2021-07-31T18:02:49.220559Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"We can now run MAGIC on the data again with the new parameters. Given that we have already fitted our MAGIC operator to the data, we should run the magic_op.transform method.","metadata":{}},{"cell_type":"code","source":"bmmsc_magic = magic_op.transform(genes=[\"Mpo\", \"Klf1\", \"Ifitm1\"])\nbmmsc_magic.head()","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:02:49.222938Z","iopub.execute_input":"2021-07-31T18:02:49.2233Z","iopub.status.idle":"2021-07-31T18:02:49.28018Z","shell.execute_reply.started":"2021-07-31T18:02:49.223251Z","shell.execute_reply":"2021-07-31T18:02:49.279096Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"fig, (ax1, ax2) = plt.subplots(1,2, figsize=(16, 6))\n\nscprep.plot.scatter(x=bmmsc_data['Mpo'], y=bmmsc_data['Klf1'], c=bmmsc_data['Ifitm1'],  ax=ax1,\n                    xlabel='Mpo', ylabel='Klf1', legend_title=\"Ifitm1\", title='Before MAGIC')\n\nscprep.plot.scatter(x=bmmsc_magic['Mpo'], y=bmmsc_magic['Klf1'], c=bmmsc_magic['Ifitm1'], ax=ax2,\n                    xlabel='Mpo', ylabel='Klf1', legend_title=\"Ifitm1\", title='After MAGIC')\n\nplt.tight_layout()\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:02:49.281381Z","iopub.execute_input":"2021-07-31T18:02:49.281656Z","iopub.status.idle":"2021-07-31T18:02:50.089086Z","shell.execute_reply.started":"2021-07-31T18:02:49.281628Z","shell.execute_reply":"2021-07-31T18:02:50.087093Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"That looks better. The gene-gene relationships are restored without smoothing so far as to remove structure.","metadata":{}},{"cell_type":"markdown","source":"## Visualizing cell trajectories with PCA on MAGIC\n\nWe can extract the principal components of the smoothed data by passing the keyword genes='pca_only' and use this for visualizing the data.","metadata":{}},{"cell_type":"code","source":"bmmsc_magic_pca = magic_op.transform(genes=\"pca_only\")\nbmmsc_magic_pca.head()","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:02:50.090461Z","iopub.execute_input":"2021-07-31T18:02:50.090882Z","iopub.status.idle":"2021-07-31T18:02:50.174412Z","shell.execute_reply.started":"2021-07-31T18:02:50.090842Z","shell.execute_reply":"2021-07-31T18:02:50.173174Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"We'll also perform PCA on the raw data for comparison.","metadata":{}},{"cell_type":"code","source":"from sklearn.decomposition import PCA\nbmmsc_pca = PCA(n_components=3).fit_transform(np.array(bmmsc_data))","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:02:50.175685Z","iopub.execute_input":"2021-07-31T18:02:50.176015Z","iopub.status.idle":"2021-07-31T18:02:51.931194Z","shell.execute_reply.started":"2021-07-31T18:02:50.175982Z","shell.execute_reply":"2021-07-31T18:02:51.93015Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"fig, (ax1, ax2) = plt.subplots(1,2, figsize=(16, 6))\n\nscprep.plot.scatter2d(bmmsc_pca, c=bmmsc_data['Ifitm1'], \n                      label_prefix=\"PC\", title='PCA without MAGIC',\n                      legend_title=\"Ifitm1\", ax=ax1, ticks=False)\n\nscprep.plot.scatter2d(bmmsc_magic_pca, c=bmmsc_magic['Ifitm1'], \n                      label_prefix=\"PC\", title='PCA with MAGIC',\n                      legend_title=\"Ifitm1\", ax=ax2, ticks=False)\n\nplt.tight_layout()\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:02:51.932677Z","iopub.execute_input":"2021-07-31T18:02:51.933029Z","iopub.status.idle":"2021-07-31T18:02:52.521666Z","shell.execute_reply.started":"2021-07-31T18:02:51.932995Z","shell.execute_reply":"2021-07-31T18:02:52.520867Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"We can also plot this in 3D.","metadata":{}},{"cell_type":"code","source":"from mpl_toolkits.mplot3d import Axes3D\n\nfig, (ax1, ax2) = plt.subplots(1,2, figsize=(16, 6), subplot_kw={'projection':'3d'})\n\nscprep.plot.scatter3d(bmmsc_pca, c=bmmsc_data['Ifitm1'], \n                      label_prefix=\"PC\", title='PCA without MAGIC',\n                      legend_title=\"Ifitm1\", ax=ax1, ticks=False)\n\nscprep.plot.scatter3d(bmmsc_magic_pca, c=bmmsc_magic['Ifitm1'], \n                      label_prefix=\"PC\", title='PCA with MAGIC',\n                      legend_title=\"Ifitm1\", ax=ax2, ticks=False)\n\nplt.tight_layout()\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:02:52.522745Z","iopub.execute_input":"2021-07-31T18:02:52.523215Z","iopub.status.idle":"2021-07-31T18:02:53.228352Z","shell.execute_reply.started":"2021-07-31T18:02:52.523171Z","shell.execute_reply":"2021-07-31T18:02:53.227052Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Visualizing MAGIC values with PHATE \n\nIn complex systems, two dimensions of PCA are not sufficient to view the entire space. For this, PHATE is a suitable visualization tool which works hand in hand with MAGIC to view how gene expression evolves along a trajectory. For this, you will need to have installed PHATE. For help using PHATE, visit https://phate.readthedocs.io/.","metadata":{}},{"cell_type":"code","source":"!pip install --user phate\n","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:02:53.230077Z","iopub.execute_input":"2021-07-31T18:02:53.230491Z","iopub.status.idle":"2021-07-31T18:03:01.235322Z","shell.execute_reply.started":"2021-07-31T18:02:53.230447Z","shell.execute_reply":"2021-07-31T18:03:01.234356Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"import phate","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:03:01.236558Z","iopub.execute_input":"2021-07-31T18:03:01.236826Z","iopub.status.idle":"2021-07-31T18:03:01.249897Z","shell.execute_reply.started":"2021-07-31T18:03:01.236797Z","shell.execute_reply":"2021-07-31T18:03:01.248953Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"data_phate = phate.PHATE().fit_transform(bmmsc_data)\n","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:03:01.251203Z","iopub.execute_input":"2021-07-31T18:03:01.251514Z","iopub.status.idle":"2021-07-31T18:03:43.711248Z","shell.execute_reply.started":"2021-07-31T18:03:01.251472Z","shell.execute_reply":"2021-07-31T18:03:43.710267Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"scprep.plot.scatter2d(data_phate, c=bmmsc_magic['Ifitm1'], figsize=(12,9),\n                      ticks=False, label_prefix=\"PHATE\", legend_title=\"Ifitm1\")","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:03:43.712555Z","iopub.execute_input":"2021-07-31T18:03:43.712836Z","iopub.status.idle":"2021-07-31T18:03:44.047325Z","shell.execute_reply.started":"2021-07-31T18:03:43.712808Z","shell.execute_reply":"2021-07-31T18:03:44.046606Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Note that the structure of the data that we see here is much more subtle than in PCA. We see multiple branches at both ends of the trajectory. To learn more about PHATE, visit https://phate.readthedocs.io/.","metadata":{}},{"cell_type":"markdown","source":"## Exact vs approximate MAGIC\nIf we are imputing many genes at once, we can speed this process up with the argument solver='approximate', which applies denoising in the PCA space and then projects these denoised principal components back onto the genes of interest. Note that this may return some small negative values. You will see below, however, that the results are largely similar to exact MAGIC.","metadata":{}},{"cell_type":"code","source":"approx_magic_op = magic.MAGIC(solver=\"approximate\")\napprox_bmmsc_magic = approx_magic_op.fit_transform(bmmsc_data, genes='all_genes')","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:03:44.048396Z","iopub.execute_input":"2021-07-31T18:03:44.048822Z","iopub.status.idle":"2021-07-31T18:03:49.818907Z","shell.execute_reply.started":"2021-07-31T18:03:44.048778Z","shell.execute_reply":"2021-07-31T18:03:49.817903Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"fig, (ax1, ax2) = plt.subplots(1,2, figsize=(16, 6))\n\nscprep.plot.scatter(x=bmmsc_magic['Mpo'], y=bmmsc_magic['Klf1'], c=bmmsc_magic['Ifitm1'],  ax=ax1,\n                    xlabel='Mpo', ylabel='Klf1', legend_title=\"Ifitm1\", title='Exact MAGIC')\n\nscprep.plot.scatter(x=approx_bmmsc_magic['Mpo'], y=approx_bmmsc_magic['Klf1'], c=approx_bmmsc_magic['Ifitm1'], ax=ax2,\n                    xlabel='Mpo', ylabel='Klf1', legend_title=\"Ifitm1\", title='Approximate MAGIC')\n\nplt.tight_layout()\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:03:49.820662Z","iopub.execute_input":"2021-07-31T18:03:49.8214Z","iopub.status.idle":"2021-07-31T18:03:50.582899Z","shell.execute_reply.started":"2021-07-31T18:03:49.82135Z","shell.execute_reply":"2021-07-31T18:03:50.581829Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Animating the MAGIC smoothing processÂ¶\nTo visualize what it means to set t in MAGIC, we can plot an animation of the smoothing process, from raw to imputed values. Below, we show an animation of Mpo, Klf1 and Ifitm1 with increasingly more smoothing.","metadata":{}},{"cell_type":"code","source":"magic.plot.animate_magic(bmmsc_data, gene_x=\"Mpo\", gene_y=\"Klf1\", gene_color=\"Ifitm1\",\n                         operator=magic_op, t_max=10)","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:03:50.587225Z","iopub.execute_input":"2021-07-31T18:03:50.587626Z","iopub.status.idle":"2021-07-31T18:03:50.839715Z","shell.execute_reply.started":"2021-07-31T18:03:50.587591Z","shell.execute_reply":"2021-07-31T18:03:50.838251Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Using the MAGIC smoothed data in downstream analysis\n\nFinally, if you wish to use the full smoothed matrix in any downstream analysis, you can extract it with the keyword genes=\"all_genes\". Note that this matrix may be very large.","metadata":{}},{"cell_type":"code","source":"bmmsc_magic = magic_op.transform(genes=\"all_genes\")\nbmmsc_magic.head()","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:10:20.838335Z","iopub.execute_input":"2021-07-31T18:10:20.838723Z","iopub.status.idle":"2021-07-31T18:10:20.885621Z","shell.execute_reply.started":"2021-07-31T18:10:20.838693Z","shell.execute_reply":"2021-07-31T18:10:20.884281Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"If you wish to export the data, you can do so in many different ways. We recommend the feather format, which is fast and lightweight, but csv is more commonly used and is compatible with many more software packages.","metadata":{}},{"cell_type":"code","source":"!pip install --user feather-format","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:03:50.843703Z","iopub.status.idle":"2021-07-31T18:03:50.844565Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"import feather\nfeather.write_dataframe(bmmsc_magic, \"bmmsc_magic.feather\")","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:03:50.845381Z","iopub.status.idle":"2021-07-31T18:03:50.845842Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"print(np.round(time.time()-t0start,1),  np.round( (time.time()-t0start)/60,1),  np.round( (time.time()-t0start)/3600,1), \n      'seconds, munutes, hours  passed')","metadata":{"execution":{"iopub.status.busy":"2021-07-31T18:03:50.846933Z","iopub.status.idle":"2021-07-31T18:03:50.847352Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]}]}