{"cells":[{"metadata":{},"cell_type":"markdown","source":"# SARS-COV-2 Gene Prediction using SARS-COV-1 Proteome\n\n## Objectives\n\n* Use known proteins associated with SARS-COV-1 to find homogologous regions in the SARS-COV-2 genome. \n"},{"metadata":{},"cell_type":"markdown","source":"## Identification of Possible Coding Regions\n\nFirst we read in the SARS-Cov-2 DNA sequence. In reality, SARS-COV-2 is a positive-sense single stranded **RNA** virus, but for some reason the data is specified in terms of DNA. The DNA sequence is then transcribed into an RNA sequence simply by replacing Ts representing thymine with Us representing Uracil. In order to translate RNA into possible amino acid sequences, we need to consider that any given strand of RNA can be read in three different ways depending on the position from which translation is started. Therefore, we consider all three reading frames when finding possible coding regions. We do this by sequentially shifting the starting position and translating the resulting reading frame into amino acids. Once each reading frame is translated, possible coding regions are identified as regions between subsequent stop codons and stored in a dataframe for later use. "},{"metadata":{"trusted":true},"cell_type":"code","source":"import numpy as np\nimport pandas as pd\nimport matplotlib.pyplot as plt\nfrom Bio import SeqIO\nfrom Bio.Alphabet import IUPAC\n\n#read in SARS-COV-1 DNA sequence\nDNAseq = SeqIO.read('../input/coronavirus-genome-sequence/MN908947.fna', \"fasta\", alphabet = IUPAC.unambiguous_dna)\nDNA = DNAseq.seq\n#transcribe sequence to rna\nmRNA = DNA.transcribe()\n\n#list to store open reading frames\nFrames = []\n\n#loop through all three possible reading frames\nfor frame in range(3):\n    #ensures reading frame is multiple of 3\n    length = 3 * ((len(mRNA)-frame) // 3)\n    #translate reading frame into amino acid sequence and split by stop codons to find all possible open reading frames\n    ORFs = mRNA[frame: frame+length].translate().split('*')\n    #append the set of open reading frames for each reading frame to list Frames\n    Frames.append(ORFs)    \n\n#list to store dfs for concatenation\nCOV19_dfs = []\n\n#loop through three possible reading frames\nfor frame in range(3):\n    #create df having for storing amino acid sequences for each ORF\n    df = pd.DataFrame(Frames[frame], columns= ['Amino Acid Sequence'])\n    #create sequence length column\n    df['length'] = df['Amino Acid Sequence'].apply(len)\n    #create columns for ORF start and stop positions \n    df['reading_frame_start_pos'] = frame + 1\n    df['reading_frame_end_pos'] = 0\n    #reading frame column\n    df['frame'] = frame\n    #calculate nucleotide start position for each ORF (not zero indexed)\n    for i in range(1,len(df)+1):\n        df['reading_frame_start_pos'].at[i] = df.iloc[i-1]['reading_frame_start_pos'] + 3*(df.iloc[i-1]['length']+1)\n    #calculate nucleotide end position for each ORF\n    for i in range(0,len(df)):\n        df['reading_frame_end_pos'].at[i]=  df.loc[i]['reading_frame_start_pos'] + 3*(df.loc[i]['length']+1)-1\n    #append df to list\n    COV19_dfs.append(df)\n#concatenate dfs in list to single df\nCOV19_df = pd.concat(COV19_dfs, ignore_index = True)\n#filter to drop amino acid sequences less than 39 which is the length of the shortest SARS-2 protein\nCOV19_df = COV19_df.loc[COV19_df.length >= 39]\n#reset index\nCOV19_df.reset_index(drop =True, inplace =True)\n\nCOV19_df.head()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"markdown","source":"## Local Alignment between SARS-COV-1 Proteins and SARS-COV-2 Coding Regions\n\nNow that we have identified promising coding regions in the SARS-Cov-1 RNA sequence, we can compare them to known SARS-COV-2 proteins. If we find a coding region which has a high degree of similarity to a known protein we can infer that the region likely encodes for a protein with a similar functionality to the known protein. We do this by performing local alignments between each protein contained in the SARS-Cov-1 proteome and each possible coding region identified above. We score each alignment and return the best match. The results are shown in the table below. "},{"metadata":{"trusted":true},"cell_type":"code","source":"#read in SARS-1 proteome\nSARS_Proteome = SeqIO.parse('../input/human-sars-coronavirus-sarscov-proteome/uniprot-proteome_UP000000354.fasta', \"fasta\")\nSARS_Proteome = [p for p in SARS_Proteome]\n# list of protein sequences\nsequence = [p.seq for p in SARS_Proteome]\n#list of protein ids\nidentity = [p.id for p in SARS_Proteome]\n#instantiate dataframe indexed by protein for storing protein amino acid sequences\nSARS_df = pd.DataFrame({'Amino Acid Sequence':sequence}, index = identity)\n#create columns for later use\nSARS_df['best_match'] = 0\nSARS_df['match_score'] = 0.0\nSARS_df['offset'] = 0\n#calcualte protein amino acid sequence lengths\nSARS_df['length'] = SARS_df['Amino Acid Sequence'].apply(len)\n\n\nfrom Bio import pairwise2\n\n#iterate over SARS-1 proteins\nfor i, r in SARS_df.iterrows():\n    #dictionaries for storing match scores and alignment offsets\n    scores = {}\n    offsets = {}\n    #for each SARS-1 protein iterate over all SARS-2 open reading frames\n    for index, row in COV19_df.iterrows():\n        #calculate local pairwise alignment between protein and ORF amino acid sequence\n        alignment = pairwise2.align.localxx(r['Amino Acid Sequence'], row['Amino Acid Sequence'], one_alignment_only = True)\n        #calculate match score and store in dictionary\n        scores[index]= alignment[0][2]/(alignment[0][4]-alignment[0][3])\n        #store offset in dictionary\n        offsets[index] = alignment[0][3]\n    #retrieve index of best match score and store value in best_match column\n    SARS_df.at[i, 'best_match'] = max(scores, key = scores.get) \n    #store best match score in match_score column\n    SARS_df.at[i,'match_score'] = max([val for k, val in scores.items()])\n    #store offset in offset column\n    SARS_df.at[i, 'offset'] = offsets[max(scores, key = scores.get)]\n#sort dataframe by match_score in descending order\nSARS_df.sort_values(by = 'match_score',ascending = False)\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### Results\n\nLet's take a look at some of the highest matching proteins:\n\n* P59637 matched coding-region 8 with a score of .91: According to [UniProt.org](https://www.uniprot.org/uniprot/P59637) this protein is an envelope small membrane protein which 'plays a central role in virus morphogenesis and assembly. Acts as a viroporin and self-assembles in host membranes forming pentameric protein-lipid pores that allow ion transport. Also plays a role in the induction of apoptosis (By similarity). Activates the host NLRP3 inflammasome, leading to IL-1beta overproduction.'\n* P59595 matched coding-region 34 with a score of .84: According to [UniProt.org](https://www.uniprot.org/uniprot/P59595), thi protein is a nucleoprotein which 'Packages the positive strand viral genome RNA into a helical ribonucleocapsid (RNP) and plays a fundamental role during virion assembly through its interactions with the viral genome and membrane protein M. Plays an important role in enhancing the efficiency of subgenomic viral RNA transcription as well as viral replication.'\n* P59596 matched coding region 74 with a score of .83: According to [UniProt.org](https://www.uniprot.org/uniprot/P59596), this protein is a membrane protein and a 'component of the viral envelope that plays a central role in virus morphogenesis and assembly via its interactions with other viral proteins.'\n* P59635 matched coding region 10 with a score of .77: According to [UniProt.org](https://www.uniprot.org/uniprot/P59635), this protein is a 'non-structural protein which is dispensable for virus replication in cell culture'.\n* P0C6U8 matched coding region 15 with a score of .7: According to [UniProt.org](https://www.uniprot.org/uniprot/P0C6U8) this protein is a polyprotein with a number of functions. \n* P59594 matched coding region 30 with a score of .65: According to [UniProt.org](https://www.uniprot.org/uniprot/P59594), this protein is a spike glycoprotein which 'attaches the virion to the cell membrane by interacting with host receptor, initiating the infection (By similarity). Binding to human ACE2 and CLEC4M/DC-SIGNR receptors and internalization of the virus into the endosomes of the host cell induces conformational changes in the S glycoprotein. Proteolysis by cathepsin CTSL may unmask the fusion peptide of S2 and activate membranes fusion within endosomes.'\n"},{"metadata":{},"cell_type":"markdown","source":"### Example Alignment for P56937 with 91% match"},{"metadata":{"trusted":true},"cell_type":"code","source":"alignment = pairwise2.align.localxx(SARS_df.loc['sp|P59637|VEMP_CVHSA', 'Amino Acid Sequence'], COV19_df.loc[8]['Amino Acid Sequence'],  penalize_extend_when_opening= True)\nprint(pairwise2.format_alignment(*alignment[0]))\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## Record Homolog Locations and Write to File\n\nFinally we create a genbank file containing the locations of the gene homologs for later use. "},{"metadata":{"trusted":true},"cell_type":"code","source":"from Bio.SeqRecord import SeqRecord\nfrom Bio.SeqFeature import SeqFeature, FeatureLocation\n\n#list for storing sequence feature\nfeatures = []\n#iterate over SARS_df by index and row\nfor i, r in SARS_df.iterrows():\n    #calculate the nucleotide start position for the region of best match in the SARS-2 genome\n    start = (COV19_df.loc[r['best_match']]['reading_frame_start_pos']-1) + r['offset']*3\n    #calcualte the nucleotide end position for the region of best match in teh SARS-2 genome\n    end = COV19_df.loc[r['best_match']]['reading_frame_end_pos']\n    #create feature\n    feature = SeqFeature(FeatureLocation(int(start), int(end)), type=\"CDS\", id = 'homolog-'+i, qualifiers = {'match_score': r['match_score'], 'homolog':i} )\n    #append to feature list\n    features.append(feature)\n#create sequenc record with features\nrecord = SeqRecord(DNA, id = 'MN908947', name = 'SARS-CoV2', description = 'potential homologous genes to SARS-CoV1', features = features)\n#write to genbank file for later use\nSeqIO.write(record, \"cov2_homologs.gb\", \"genbank\")","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## Conclusion\n\nSince we found several coding regions which exhibited a high degree of similarity to known proteins in the SARS-Cov-1 proteome, we can infer that these coding regions likely encode for proteins having similar functions to those matched. "}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}